{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Conservation Analysis and Epitope Prediction\n",
    "\n",
    "\n",
    "#### Author: C. Mazzaferro, K. Fisch\n",
    "#### Email: cmazzafe@ucsd.edu\n",
    "#### Date: October 2016\n",
    " \n",
    "## Outline of Notebook\n",
    "<a id = \"toc\"></a>\n",
    "1. <a href = \"#background\">Background</a>\n",
    "2. <a href = \"#Cons\">High Affinity Binding Prediction </a>\n",
    "    * <a href = \"#Agg\">Data Aggregation</a>\n",
    "    * <a href = \"#Sim\">Similarty Score Calculation</a>\n",
    "    * <a href = \"#Plot\">Visualize</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Files of interest\n",
    "From a simple fasta file containing the proteins of interest, this workflow will create:\n",
    "\n",
    "1. Binding affinity prediction xls files for the given allele, nmer combinations. Default set to all allele supertypes and [8, 9, 10, 11] nmers\n",
    "2. Csv file containing all high binding affinity peptides. Threshold can be set to a customized level (either by setting it to Medium or setting it by affinity score. See [CBS's prediction services](http://www.cbs.dtu.dk/services/), in particular [netMHC](http://www.cbs.dtu.dk/services/NetMHC/) for more specific information.\n",
    "3. Fasta files containing swap data: for each high affinity binding peptide, find every peptide that results from swapping one single AA from it. These fasta files are then fed into netMHC again and their binding affinity determined again. \n",
    "4. Csv files for each fasta file with the swapped peptide binding prediction.\n",
    "\n",
    "Caveats: netMHC local install must be in the system's path. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import importlib\n",
    "from nepitope import net_MHC_interface\n",
    "from nepitope import mhc_utils\n",
    "from nepitope import pep_utils\n",
    "import pandas\n",
    "\n",
    "project_location = '/Users/carlomazzaferro/Desktop/New_General/'\n",
    "fasta_file = '/Users/carlomazzaferro/Desktop/New_General/fasta_base_new_prots.fasta'\n",
    "net_mhc_path = '/Users/carlomazzaferro/Desktop/BINF_Tools/netMHC-4.0/netMHC'\n",
    "\n",
    "\"\"\"\n",
    "Default Parameters:\n",
    "\n",
    "all_alleles = ['HLA-A0101','HLA-A0201','HLA-A0301','HLA-A2402','HLA-A2601',\n",
    "               'HLA-B0702','HLA-B0801','HLA-B2705','HLA-B3901','HLA-B4001','HLA-B5801','HLA-B1501']\n",
    "\n",
    "nmer_list = [8,9,10,11]\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "net_mhc = net_MHC_interface.netMHCComand(net_mhc_path, fasta_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id = \"Cons\"></a>\n",
    "# MHC Binding Affinity Prediction \n",
    "### Ran on netMHCcons for n-mers of 8, 9, 10, 11 amino acids and for the 12 allele supertype families\n",
    "\n",
    "Run for each protein sequence retrieved from reference paper."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'netMHC commands written to run_netMHC.txt located at /Users/carlomazzaferro/Desktop/New_General/mhc_preds_fasta_base_new_prots'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "net_mhc.create_text_command()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/mhc_preds_fasta_base_new_prots\n"
     ]
    }
   ],
   "source": [
    "net_mhc.run_netMHC()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Load predictions \n",
    "import importlib\n",
    "importlib.reload(mhc_utils)\n",
    "pred_loc = '/Users/carlomazzaferro/Desktop/New_General/mhc_preds_fasta_base_new_prots/'\n",
    "agg = mhc_utils.FileConsolidation(pred_loc, fasta_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/carlomazzaferro/anaconda/envs/py35/lib/python3.5/site-packages/pandas/core/generic.py:2773: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self[name] = value\n"
     ]
    }
   ],
   "source": [
    "csv_dir = '/Users/carlomazzaferro/Desktop/New_General/high_affinity_peps.csv'\n",
    "high_aa_df = agg.get_all_high_affinity_from_batch(csv_out=True, csv_dir=csv_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pos</th>\n",
       "      <th>Peptide</th>\n",
       "      <th>nM</th>\n",
       "      <th>Rank</th>\n",
       "      <th>ID</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Affinity Level</th>\n",
       "      <th>n-mer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1003</th>\n",
       "      <td>1003</td>\n",
       "      <td>LESEFVYGDY</td>\n",
       "      <td>33.5</td>\n",
       "      <td>0.06</td>\n",
       "      <td>S__pyogenes_Cas9</td>\n",
       "      <td>HLA-A0101</td>\n",
       "      <td>High</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2388</th>\n",
       "      <td>1029</td>\n",
       "      <td>YSTDILGNLY</td>\n",
       "      <td>3.7</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Staphylococcus_aureus_Cas9</td>\n",
       "      <td>HLA-A0101</td>\n",
       "      <td>High</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389</th>\n",
       "      <td>1030</td>\n",
       "      <td>STDILGNLYE</td>\n",
       "      <td>47.1</td>\n",
       "      <td>0.08</td>\n",
       "      <td>Staphylococcus_aureus_Cas9</td>\n",
       "      <td>HLA-A0101</td>\n",
       "      <td>High</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2608</th>\n",
       "      <td>205</td>\n",
       "      <td>QITDEFINRY</td>\n",
       "      <td>44.0</td>\n",
       "      <td>0.08</td>\n",
       "      <td>S_CRISPR_1_thermophilus_Cas9</td>\n",
       "      <td>HLA-A0101</td>\n",
       "      <td>High</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7793</th>\n",
       "      <td>1102</td>\n",
       "      <td>VRTDVLGNKY</td>\n",
       "      <td>38.1</td>\n",
       "      <td>0.07</td>\n",
       "      <td>Streptococcus_pasteurianus_Cas9</td>\n",
       "      <td>HLA-A0101</td>\n",
       "      <td>High</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Pos     Peptide    nM  Rank                               ID  \\\n",
       "1003  1003  LESEFVYGDY  33.5  0.06                 S__pyogenes_Cas9   \n",
       "2388  1029  YSTDILGNLY   3.7  0.01       Staphylococcus_aureus_Cas9   \n",
       "2389  1030  STDILGNLYE  47.1  0.08       Staphylococcus_aureus_Cas9   \n",
       "2608   205  QITDEFINRY  44.0  0.08     S_CRISPR_1_thermophilus_Cas9   \n",
       "7793  1102  VRTDVLGNKY  38.1  0.07  Streptococcus_pasteurianus_Cas9   \n",
       "\n",
       "         Allele Affinity Level  n-mer  \n",
       "1003  HLA-A0101           High     10  \n",
       "2388  HLA-A0101           High     10  \n",
       "2389  HLA-A0101           High     10  \n",
       "2608  HLA-A0101           High     10  \n",
       "7793  HLA-A0101           High     10  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "high_aa_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Swap Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Peptide</th>\n",
       "      <th>nM</th>\n",
       "      <th>Rank</th>\n",
       "      <th>ID</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Affinity Level</th>\n",
       "      <th>n-mer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1003</td>\n",
       "      <td>1003</td>\n",
       "      <td>LESEFVYGDY</td>\n",
       "      <td>33.5</td>\n",
       "      <td>0.06</td>\n",
       "      <td>S__pyogenes_Cas9</td>\n",
       "      <td>HLA-A0101</td>\n",
       "      <td>High</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2388</td>\n",
       "      <td>1029</td>\n",
       "      <td>YSTDILGNLY</td>\n",
       "      <td>3.7</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Staphylococcus_aureus_Cas9</td>\n",
       "      <td>HLA-A0101</td>\n",
       "      <td>High</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2389</td>\n",
       "      <td>1030</td>\n",
       "      <td>STDILGNLYE</td>\n",
       "      <td>47.1</td>\n",
       "      <td>0.08</td>\n",
       "      <td>Staphylococcus_aureus_Cas9</td>\n",
       "      <td>HLA-A0101</td>\n",
       "      <td>High</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2608</td>\n",
       "      <td>205</td>\n",
       "      <td>QITDEFINRY</td>\n",
       "      <td>44.0</td>\n",
       "      <td>0.08</td>\n",
       "      <td>S_CRISPR_1_thermophilus_Cas9</td>\n",
       "      <td>HLA-A0101</td>\n",
       "      <td>High</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7793</td>\n",
       "      <td>1102</td>\n",
       "      <td>VRTDVLGNKY</td>\n",
       "      <td>38.1</td>\n",
       "      <td>0.07</td>\n",
       "      <td>Streptococcus_pasteurianus_Cas9</td>\n",
       "      <td>HLA-A0101</td>\n",
       "      <td>High</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0   Pos     Peptide    nM  Rank                               ID  \\\n",
       "0        1003  1003  LESEFVYGDY  33.5  0.06                 S__pyogenes_Cas9   \n",
       "1        2388  1029  YSTDILGNLY   3.7  0.01       Staphylococcus_aureus_Cas9   \n",
       "2        2389  1030  STDILGNLYE  47.1  0.08       Staphylococcus_aureus_Cas9   \n",
       "3        2608   205  QITDEFINRY  44.0  0.08     S_CRISPR_1_thermophilus_Cas9   \n",
       "4        7793  1102  VRTDVLGNKY  38.1  0.07  Streptococcus_pasteurianus_Cas9   \n",
       "\n",
       "      Allele Affinity Level  n-mer  \n",
       "0  HLA-A0101           High     10  \n",
       "1  HLA-A0101           High     10  \n",
       "2  HLA-A0101           High     10  \n",
       "3  HLA-A0101           High     10  \n",
       "4  HLA-A0101           High     10  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "high_aa_df = pandas.read_csv(csv_dir) \n",
    "high_aa_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from nepitope import peptide_utilities\n",
    "importlib.reload(peptide_utilities)\n",
    "fasta_files_dir = '/Users/carlomazzaferro/Desktop/New_General/swaps/'\n",
    "\n",
    "swap = peptide_utilities.Swaps(high_aa_df, fasta_files_dir, net_mhc_path, proteins=['S__pyogenes_Cas9'] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_LESEFVYGDY_Pos_1003_ID_S--pyogenes-Cas9_Allele_HLA-A0101_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_NGLFGNLIAL_Pos_234_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_AILLSDILRV_Pos_298_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_ILTFRIPYYV_Pos_442_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_HSLLYEYFTV_Pos_510_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_YLQNGRDMYV_Pos_814_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_YLNAVVGTAL_Pos_987_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_KLPKYSLFEL_Pos_1196_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_HYLDEIIEQI_Pos_1263_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_KAILSARLSK_Pos_208_ID_S--pyogenes-Cas9_Allele_HLA-A0301_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_KLNREDLLRK_Pos_391_ID_S--pyogenes-Cas9_Allele_HLA-A0301_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_KVLPKHSLLY_Pos_505_ID_S--pyogenes-Cas9_Allele_HLA-A0301_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_SLGTYHDLLK_Pos_589_ID_S--pyogenes-Cas9_Allele_HLA-A0301_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_KTYAHLFDDK_Pos_636_ID_S--pyogenes-Cas9_Allele_HLA-A0301_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_TIMERSSFEK_Pos_1166_ID_S--pyogenes-Cas9_Allele_HLA-A0301_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_MLASAGELQK_Pos_1212_ID_S--pyogenes-Cas9_Allele_HLA-A0301_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_FLYLASHYEK_Pos_1234_ID_S--pyogenes-Cas9_Allele_HLA-A0301_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_TNLGAPAAFK_Pos_1315_ID_S--pyogenes-Cas9_Allele_HLA-A0301_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_EEVVKKMKNY_Pos_872_ID_S--pyogenes-Cas9_Allele_HLA-A2601_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_EFVYGDYKVY_Pos_1006_ID_S--pyogenes-Cas9_Allele_HLA-A2601_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_DEIIEQISEF_Pos_1266_ID_S--pyogenes-Cas9_Allele_HLA-A2601_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_EIIEQISEFS_Pos_1267_ID_S--pyogenes-Cas9_Allele_HLA-A2601_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_IPHQIHLGEL_Pos_409_ID_S--pyogenes-Cas9_Allele_HLA-B0702_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_HMIKFRGHFL_Pos_159_ID_S--pyogenes-Cas9_Allele_HLA-B0801_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_RLIYLALAHM_Pos_151_ID_S--pyogenes-Cas9_Allele_HLA-B1501_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_AHMIKFRGHF_Pos_158_ID_S--pyogenes-Cas9_Allele_HLA-B1501_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_KLFIQLVQTY_Pos_182_ID_S--pyogenes-Cas9_Allele_HLA-B1501_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_ALSLGLTPNF_Pos_242_ID_S--pyogenes-Cas9_Allele_HLA-B1501_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_SLLYEYFTVY_Pos_511_ID_S--pyogenes-Cas9_Allele_HLA-B1501_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_KQSGKTILDF_Pos_672_ID_S--pyogenes-Cas9_Allele_HLA-B1501_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_QLIHDDSLTF_Pos_694_ID_S--pyogenes-Cas9_Allele_HLA-B1501_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_LQNEKLYLYY_Pos_805_ID_S--pyogenes-Cas9_Allele_HLA-B1501_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_LIHQSITGLY_Pos_1346_ID_S--pyogenes-Cas9_Allele_HLA-B1501_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_RRYTRRKNRI_Pos_69_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_RRLENLIAQL_Pos_219_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_RRQEDFYPFL_Pos_423_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_NSRFAWMTRK_Pos_458_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_SRFAWMTRKS_Pos_459_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_RRRYTGWGRL_Pos_652_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_RRYTGWGRLS_Pos_653_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_WRQLLNAKLI_Pos_882_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_KRMLASAGEL_Pos_1210_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_YEYFTVYNEL_Pos_514_ID_S--pyogenes-Cas9_Allele_HLA-B4001_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_GVEDRFNASL_Pos_581_ID_S--pyogenes-Cas9_Allele_HLA-B4001_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_REQAENIIHL_Pos_1302_ID_S--pyogenes-Cas9_Allele_HLA-B4001_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_AENIIHLFTL_Pos_1305_ID_S--pyogenes-Cas9_Allele_HLA-B4001_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_VSDFRKDFQFY_Pos_962_ID_S--pyogenes-Cas9_Allele_HLA-A0101_nmer_11\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_KVYDVRKMIAK_Pos_1013_ID_S--pyogenes-Cas9_Allele_HLA-A0301_nmer_11\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_RRLENLIAQLP_Pos_219_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_11\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_LLYEYFTV_Pos_512_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_8\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_MIKFRGHF_Pos_160_ID_S--pyogenes-Cas9_Allele_HLA-B0801_nmer_8\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_RRYTGWGR_Pos_653_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_8\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_ESEFVYGDY_Pos_1004_ID_S--pyogenes-Cas9_Allele_HLA-A0101_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_NLDKVLSAY_Pos_1285_ID_S--pyogenes-Cas9_Allele_HLA-A0101_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_GLDIGTNSV_Pos_7_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_GLFGNLIAL_Pos_235_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_NLIALSLGL_Pos_239_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_ILLSDILRV_Pos_299_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_LTFRIPYYV_Pos_443_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_SLLYEYFTV_Pos_511_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_SLHEHIANL_Pos_718_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_YLASHYEKL_Pos_1236_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_YLDEIIEQI_Pos_1264_ID_S--pyogenes-Cas9_Allele_HLA-A0201_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_AILSARLSK_Pos_209_ID_S--pyogenes-Cas9_Allele_HLA-A0301_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_KTNRKVTVK_Pos_553_ID_S--pyogenes-Cas9_Allele_HLA-A0301_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_AYHEKYPTI_Pos_126_ID_S--pyogenes-Cas9_Allele_HLA-A2402_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_IYLALAHMI_Pos_153_ID_S--pyogenes-Cas9_Allele_HLA-A2402_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_FYSNIMNFF_Pos_1037_ID_S--pyogenes-Cas9_Allele_HLA-A2402_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_EVVKKMKNY_Pos_873_ID_S--pyogenes-Cas9_Allele_HLA-A2601_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_FVYGDYKVY_Pos_1007_ID_S--pyogenes-Cas9_Allele_HLA-A2601_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_EIIEQISEF_Pos_1267_ID_S--pyogenes-Cas9_Allele_HLA-A2601_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_VPSKKFKVL_Pos_26_ID_S--pyogenes-Cas9_Allele_HLA-B0702_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_SPAIKKGIL_Pos_729_ID_S--pyogenes-Cas9_Allele_HLA-B0702_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_SPTVAYSVL_Pos_1135_ID_S--pyogenes-Cas9_Allele_HLA-B0702_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_LPKYSLFEL_Pos_1197_ID_S--pyogenes-Cas9_Allele_HLA-B0702_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_VPSKKFKVL_Pos_26_ID_S--pyogenes-Cas9_Allele_HLA-B0801_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_MIKFRGHFL_Pos_160_ID_S--pyogenes-Cas9_Allele_HLA-B0801_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_QLKRRRYTG_Pos_649_ID_S--pyogenes-Cas9_Allele_HLA-B0801_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_FANRNFMQL_Pos_687_ID_S--pyogenes-Cas9_Allele_HLA-B0801_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_HMIKFRGHF_Pos_159_ID_S--pyogenes-Cas9_Allele_HLA-B1501_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_LLYEYFTVY_Pos_512_ID_S--pyogenes-Cas9_Allele_HLA-B1501_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_LIHDDSLTF_Pos_695_ID_S--pyogenes-Cas9_Allele_HLA-B1501_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_RRYTRRKNR_Pos_69_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_RRKNRICYL_Pos_73_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_SRFAWMTRK_Pos_459_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_RRRYTGWGR_Pos_652_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_RRYTGWGRL_Pos_653_ID_S--pyogenes-Cas9_Allele_HLA-B2705_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_MKNYWRQLL_Pos_878_ID_S--pyogenes-Cas9_Allele_HLA-B3901_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_YHHAHDAYL_Pos_980_ID_S--pyogenes-Cas9_Allele_HLA-B3901_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_AHDAYLNAV_Pos_983_ID_S--pyogenes-Cas9_Allele_HLA-B3901_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_GETAEATRL_Pos_55_ID_S--pyogenes-Cas9_Allele_HLA-B4001_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_VEDRFNASL_Pos_582_ID_S--pyogenes-Cas9_Allele_HLA-B4001_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_AERGGLSEL_Pos_902_ID_S--pyogenes-Cas9_Allele_HLA-B4001_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_GELQKGNEL_Pos_1217_ID_S--pyogenes-Cas9_Allele_HLA-B4001_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_SEFSKRVIL_Pos_1273_ID_S--pyogenes-Cas9_Allele_HLA-B4001_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_KSEETITPW_Pos_467_ID_S--pyogenes-Cas9_Allele_HLA-B5801_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_KATAKYFFY_Pos_1030_ID_S--pyogenes-Cas9_Allele_HLA-B5801_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/mhc_preds__swap_KSVKELLGI_Pos_1157_ID_S--pyogenes-Cas9_Allele_HLA-B5801_nmer_9\n"
     ]
    }
   ],
   "source": [
    "cmd = swap.find_swaps_write_to_fasta()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from nepitope import mhc_utils\n",
    "import pandas\n",
    "import glob\n",
    "import importlib\n",
    "importlib.reload(mhc_utils)\n",
    "\n",
    "#files saved from netMHC in  multiple different files, one per each query (nmer, allele)\n",
    "filepath = '/Users/carlomazzaferro/Desktop/New_General/swaps/S__pyogenes_Cas9/preds_per_swap/'\n",
    "fasta_file = '/Users/carlomazzaferro/Desktop/New_General/fasta_base_new_prots.fasta'\n",
    "\n",
    "aggregate_all = mhc_utils.FileConsolidation(filepath, fasta_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#returns a dataframe per file used (just one in this case)\n",
    "df_list = aggregate_all.return_df_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Would run this. But too slow. The 3 cells below are speed-optimized. \n",
    "lsss_1 = aggregate_all.optimized_list_df_by_prot(df_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 646,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sss = pandas.concat(lsss_1)\n",
    "high_int_aa = sss.loc[sss['Affinity Level'] ==  ('High' or 'Intermediate')]\n",
    "high_int_aa.to_csv('/Users/carlomazzaferro/Desktop/Test_Workflow_From_Scratch/all_high_aff_aas.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load original proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 514,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'nepitope.pairwise_comp' from '/Users/carlomazzaferro/Documents/Code/neoantigen/antigen_discovery/nepitope/pairwise_comp.py'>"
      ]
     },
     "execution_count": 514,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from nepitope import pairwise_comp\n",
    "importlib.reload(pairwise_comp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/carlomazzaferro/Documents/Code/neoantigen/antigen_discovery/nepitope/pairwise_comp.py:194: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  df['Peptide'] = df['Peptide'].str.replace('X', '-')\n"
     ]
    }
   ],
   "source": [
    "pwcomp = pairwise_comp.PairwiseComp(lsss_1, 5, fasta_file) \n",
    "df_pairwise_comps  = pwcomp.pipe_run() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "swaps_df_dir = '/Users/carlomazzaferro/Desktop/Test_IEDB/OtherProtsPrashant/swap_results/S__pyogenes_Cas9/summary_results_per_prediction.csv'\n",
    "swaps_df = pandas.read_csv(swaps_df_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/carlomazzaferro/Documents/Code/neoantigen/antigen_discovery/nepitope/merge_workflows.py:120: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  df = pandas.DataFrame(maximal_list)\n"
     ]
    }
   ],
   "source": [
    "from importlib import reload\n",
    "reload(merge_workflows)\n",
    "from nepitope import merge_workflows\n",
    "merged = merge_workflows.MergeSwapsAndComp(df_pairwise_comps, swaps_df, 40)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 577,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "130 matches occurring at position 657\n",
      "129 matches occurring at position 447\n",
      "126 matches occurring at position 656\n",
      "124 matches occurring at position 446\n",
      "122 matches occurring at position 658\n",
      "119 matches occurring at position 448\n",
      "117 matches occurring at position 813\n",
      "115 matches occurring at position 814\n",
      "110 matches occurring at position 659\n",
      "109 matches occurring at position 445\n",
      "109 matches occurring at position 812\n",
      "106 matches occurring at position 514\n",
      "106 matches occurring at position 655\n",
      "103 matches occurring at position 513\n",
      "100 matches occurring at position 449\n",
      "98 matches occurring at position 512\n",
      "98 matches occurring at position 515\n",
      "98 matches occurring at position 985\n",
      "98 matches occurring at position 987\n",
      "96 matches occurring at position 986\n"
     ]
    }
   ],
   "source": [
    "merged.top_loc_matches[0:20]\n",
    "for i in merged.top_loc_matches[0:20]:\n",
    "    print('%i matches occurring at position %i' %(i[1], i[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/carlomazzaferro/Documents/Code/neoantigen/antigen_discovery/nepitope/merge_workflows.py:96: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "df_top = merged.top_swap_df\n",
    "df_ordered_filtered = merged.final_sort_and_filtering(df_top)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 642,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'nmer', 'allele', 'num high affinity peps',\n",
       "       'num med affinity peps', 'num low affinity peps',\n",
       "       'num no affinity peps', 'original peptide', 'original pos',\n",
       "       'top scoring peptides', 'Range'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 642,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ordered_filtered.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 610,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "reload(modify_data)\n",
    "from nepitope import modify_data\n",
    "mod_fasta = '/Users/carlomazzaferro/Desktop/PROJECT_LOC_TEST/mod_fastas/mod_all_fastas/'\n",
    "orig_fasta = '/Users/carlomazzaferro/Desktop/PROJECT_LOC_TEST/prots_in_fasta.fasta'\n",
    "\n",
    "mod_data = modify_data.ModifyData(swaps_df, fasta_file, mod_fasta, lsss_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 611,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "True\n",
      "RRYTRRKNR\n",
      "True\n",
      "KAILSARLSK\n",
      "True\n",
      "KLNREDLLRK\n",
      "True\n",
      "WRQLLNAKLI\n",
      "True\n",
      "SRFAWMTRK\n",
      "True\n",
      "LIHDDSLTF\n",
      "True\n",
      "IYLALAHMI\n",
      "True\n",
      "YHHAHDAYL\n",
      "True\n",
      "SLLYEYFTVY\n",
      "True\n",
      "ILLSDILRV\n",
      "True\n",
      "MLASAGELQK\n",
      "True\n",
      "KQSGKTILDF\n",
      "True\n",
      "RRLENLIAQL\n",
      "True\n",
      "KTYAHLFDDK\n",
      "True\n",
      "AHMIKFRGHF\n",
      "True\n",
      "LPKYSLFEL\n",
      "True\n",
      "SLHEHIANL\n",
      "True\n",
      "SPAIKKGIL\n",
      "True\n",
      "RLIYLALAHM\n",
      "True\n",
      "HYLDEIIEQI\n",
      "True\n",
      "RRKNRICYL\n",
      "True\n",
      "REQAENIIHL\n",
      "True\n",
      "KVLPKHSLLY\n",
      "True\n",
      "MIKFRGHFL\n",
      "True\n",
      "LLYEYFTVY\n",
      "True\n",
      "ALSLGLTPNF\n",
      "True\n",
      "NGLFGNLIAL\n",
      "True\n",
      "RRYTGWGRLS\n",
      "True\n",
      "LLYEYFTV\n",
      "True\n",
      "AHDAYLNAV\n",
      "True\n",
      "YLQNGRDMYV\n",
      "True\n",
      "TIMERSSFEK\n",
      "True\n",
      "RRYTRRKNRI\n",
      "True\n",
      "ILTFRIPYYV\n",
      "True\n",
      "RRRYTGWGR\n",
      "True\n",
      "SEFSKRVIL\n",
      "True\n",
      "GLDIGTNSV\n",
      "True\n",
      "MKNYWRQLL\n",
      "True\n",
      "RRRYTGWGRL\n",
      "True\n",
      "VEDRFNASL\n",
      "True\n",
      "FVYGDYKVY\n",
      "True\n",
      "SPTVAYSVL\n",
      "True\n",
      "IPHQIHLGEL\n",
      "True\n",
      "AYHEKYPTI\n",
      "True\n",
      "FANRNFMQL\n",
      "True\n",
      "KLPKYSLFEL\n",
      "True\n",
      "LTFRIPYYV\n",
      "True\n",
      "AENIIHLFTL\n",
      "True\n",
      "AILSARLSK\n",
      "True\n",
      "HMIKFRGHF\n",
      "True\n",
      "NSRFAWMTRK\n",
      "True\n",
      "EIIEQISEF\n",
      "True\n",
      "NLIALSLGL\n",
      "True\n",
      "FLYLASHYEK\n",
      "True\n",
      "AERGGLSEL\n",
      "True\n",
      "GVEDRFNASL\n",
      "True\n",
      "DEIIEQISEF\n",
      "True\n",
      "GELQKGNEL\n",
      "True\n",
      "EEVVKKMKNY\n",
      "True\n",
      "YLDEIIEQI\n",
      "True\n",
      "GLFGNLIAL\n",
      "True\n",
      "YEYFTVYNEL\n",
      "True\n",
      "VPSKKFKVL\n",
      "True\n",
      "EFVYGDYKVY\n",
      "True\n",
      "RRLENLIAQLP\n",
      "True\n",
      "RRQEDFYPFL\n",
      "True\n",
      "QLKRRRYTG\n",
      "True\n",
      "KSVKELLGI\n",
      "True\n",
      "SRFAWMTRKS\n",
      "True\n",
      "LQNEKLYLYY\n",
      "True\n",
      "LESEFVYGDY\n",
      "True\n",
      "RRYTGWGRL\n",
      "True\n",
      "MIKFRGHF\n",
      "True\n",
      "KTNRKVTVK\n",
      "True\n",
      "FYSNIMNFF\n",
      "True\n",
      "GETAEATRL\n",
      "True\n",
      "KVYDVRKMIAK\n",
      "True\n",
      "KRMLASAGEL\n",
      "True\n",
      "VSDFRKDFQFY\n",
      "True\n",
      "EIIEQISEFS\n",
      "True\n",
      "HSLLYEYFTV\n",
      "True\n",
      "YLASHYEKL\n",
      "True\n",
      "RRYTGWGR\n",
      "True\n",
      "SLLYEYFTV\n",
      "True\n",
      "TNLGAPAAFK\n",
      "True\n",
      "QLIHDDSLTF\n",
      "True\n",
      "KSEETITPW\n",
      "True\n",
      "NLDKVLSAY\n",
      "True\n",
      "KATAKYFFY\n",
      "True\n",
      "ESEFVYGDY\n",
      "True\n",
      "EVVKKMKNY\n",
      "True\n",
      "YLNAVVGTAL\n",
      "True\n",
      "SLGTYHDLLK\n",
      "True\n",
      "LIHQSITGLY\n",
      "True\n",
      "HMIKFRGHFL\n",
      "True\n",
      "KLFIQLVQTY\n",
      "True\n",
      "AILLSDILRV\n"
     ]
    }
   ],
   "source": [
    "######## TO INCLUDE IN TEST SUITE ########\n",
    "\n",
    "from nepitope import pep_utils\n",
    "j,k = pep_utils.create_separate_lists(fasta_file)\n",
    "for i in sgl:\n",
    "    print(i[1] in k[0])\n",
    "\n",
    "for i in mod_data.swaps_dic.keys():\n",
    "    print(i in mod_data.fasta_dic[0][1])\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 612,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mod_data.get_modified_fastas_single_swap()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 615,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exchange_pairs = []\n",
    "for i,j in mod_data.swaps_dic.items():\n",
    "    exchange_pairs.append([i,j])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 603,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "def get_all_exchange_pairs(swaps_df):\n",
    "    cols = ['original peptide', 'top scoring peptides']\n",
    "    sliced = swaps_df.ix[:, cols].values.tolist()\n",
    "    orig = [i[0] for i in sliced]\n",
    "    swaps = [list(filter(None, re.sub('[^A-Za-z0-9]+', ',', i[1]).split(','))) for i in sliced]\n",
    "    hashmap = [[orig[i], swaps[i]] for i in range(len(orig))]\n",
    "    swaps_single = []\n",
    "    for idx, hsh in enumerate(hashmap):\n",
    "        swaps_single.append([hsh[0], random.choice(hsh[1])])\n",
    "    return swaps_single\n",
    "\n",
    "sgl = get_all_exchange_pairs(swaps_df)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 616,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/carlomazzaferro/Documents/Code/neoantigen/antigen_discovery/nepitope/pairwise_comp.py:194: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  df['Peptide'] = df['Peptide'].str.replace('X', '-')\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "fastas = glob.glob(mod_fasta + '*')\n",
    "csv_path = '/Users/carlomazzaferro/Desktop/PROJECT_LOC_TEST/mod_csvs/mod_csvs_all_peps/'\n",
    "\n",
    "list_dfs = []\n",
    "for pair in exchange_pairs:\n",
    "    for fasta in fastas:\n",
    "        if os.path.splitext(os.path.basename(fasta))[0].split('-')[0] == pair[0]:\n",
    "            my_fasta = fasta\n",
    "            ls_mod = mod_data.get_singly_modified_df_list(pair[0], pair[1])\n",
    "            pwcomp_pair_swap = pairwise_comp.PairwiseComp(ls_mod, 5, my_fasta)\n",
    "            df_swap = pwcomp_pair_swap.pipe_run()\n",
    "            df_swap = df_swap.drop('Matches Loc', 1)\n",
    "            df_swap.to_csv(csv_path + pair[0] + '-' + pair[1] + '.csv')\n",
    "            list_dfs.append(df_swap)\n",
    "            \n",
    "            \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 629,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>Num High AA</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C__jejuni_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Campylobacter_lari_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Corynebacter_diphtheria_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F__novicida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lactobacillus_buchneri_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Listeria_innocua_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>41</td>\n",
       "      <td>23</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N__meningitidis_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neisseria_cinerea_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P__multocida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Parvibaculum_lavamentivorans_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_CRISPR_1_thermophilus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_CRISPR_3_thermophilus_Cas9</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>169</td>\n",
       "      <td>101</td>\n",
       "      <td>59</td>\n",
       "      <td>30</td>\n",
       "      <td>13</td>\n",
       "      <td>70</td>\n",
       "      <td>374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__mutans_Cas9</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>128</td>\n",
       "      <td>88</td>\n",
       "      <td>56</td>\n",
       "      <td>30</td>\n",
       "      <td>13</td>\n",
       "      <td>87</td>\n",
       "      <td>316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__pyogenes_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27</td>\n",
       "      <td>20</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Staphylococcus_aureus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus_pasteurianus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T__denticola_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  10   11    5    6   7   8   9 Num High AA  \\\n",
       "C__jejuni_Cas9                     0    0    0    0   0   0   0           0   \n",
       "Campylobacter_lari_Cas9            0    0    0    0   0   0   0           0   \n",
       "Corynebacter_diphtheria_Cas9       0    0    0    0   0   0   0           0   \n",
       "F__novicida_Cas9                   0    0    0    0   0   0   0           0   \n",
       "Lactobacillus_buchneri_Cas9        0    0   15    5   2   1   0           2   \n",
       "Listeria_innocua_Cas9              0    0   67   41  23  13   5           7   \n",
       "N__meningitidis_Cas9               0    0    2    0   0   0   0           0   \n",
       "Neisseria_cinerea_Cas9             0    0    2    0   0   0   0           0   \n",
       "P__multocida_Cas9                  0    0    2    1   0   0   0           0   \n",
       "Parvibaculum_lavamentivorans_Cas9  0    0    0    0   0   0   0           0   \n",
       "S_CRISPR_1_thermophilus_Cas9       0    0    1    0   0   0   0           0   \n",
       "S_CRISPR_3_thermophilus_Cas9       2    0  169  101  59  30  13          70   \n",
       "S__mutans_Cas9                     1    0  128   88  56  30  13          87   \n",
       "S__pyogenes_Cas9                   0  NaN   27   20  13   6   2          68   \n",
       "Staphylococcus_aureus_Cas9         0    0    6    0   0   0   0           0   \n",
       "Streptococcus_pasteurianus_Cas9    0    0    2    0   0   0   0           0   \n",
       "T__denticola_Cas9                  0    0   27   12   2   1   0           1   \n",
       "\n",
       "                                  Total  \n",
       "C__jejuni_Cas9                        0  \n",
       "Campylobacter_lari_Cas9               0  \n",
       "Corynebacter_diphtheria_Cas9          0  \n",
       "F__novicida_Cas9                      0  \n",
       "Lactobacillus_buchneri_Cas9          23  \n",
       "Listeria_innocua_Cas9               149  \n",
       "N__meningitidis_Cas9                  2  \n",
       "Neisseria_cinerea_Cas9                2  \n",
       "P__multocida_Cas9                     3  \n",
       "Parvibaculum_lavamentivorans_Cas9     0  \n",
       "S_CRISPR_1_thermophilus_Cas9          1  \n",
       "S_CRISPR_3_thermophilus_Cas9        374  \n",
       "S__mutans_Cas9                      316  \n",
       "S__pyogenes_Cas9                     68  \n",
       "Staphylococcus_aureus_Cas9            6  \n",
       "Streptococcus_pasteurianus_Cas9       2  \n",
       "T__denticola_Cas9                    42  "
      ]
     },
     "execution_count": 629,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_dfs[97].loc['S__pyogenes_Cas9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 630,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>Num High AA</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C__jejuni_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Campylobacter_lari_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Corynebacter_diphtheria_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F__novicida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lactobacillus_buchneri_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Listeria_innocua_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>41</td>\n",
       "      <td>23</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N__meningitidis_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neisseria_cinerea_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P__multocida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Parvibaculum_lavamentivorans_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_CRISPR_1_thermophilus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_CRISPR_3_thermophilus_Cas9</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>170</td>\n",
       "      <td>101</td>\n",
       "      <td>59</td>\n",
       "      <td>30</td>\n",
       "      <td>13</td>\n",
       "      <td>86</td>\n",
       "      <td>375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__mutans_Cas9</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>131</td>\n",
       "      <td>91</td>\n",
       "      <td>59</td>\n",
       "      <td>33</td>\n",
       "      <td>15</td>\n",
       "      <td>124</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__pyogenes_Cas9</th>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>130</td>\n",
       "      <td>104</td>\n",
       "      <td>79</td>\n",
       "      <td>54</td>\n",
       "      <td>31</td>\n",
       "      <td>408</td>\n",
       "      <td>408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Staphylococcus_aureus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus_pasteurianus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T__denticola_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   10   11    5    6   7   8   9 Num High AA  \\\n",
       "C__jejuni_Cas9                      0    0    0    0   0   0   0           0   \n",
       "Campylobacter_lari_Cas9             0    0    0    0   0   0   0           0   \n",
       "Corynebacter_diphtheria_Cas9        0    0    0    0   0   0   0           0   \n",
       "F__novicida_Cas9                    0    0    0    0   0   0   0           0   \n",
       "Lactobacillus_buchneri_Cas9         0    0   15    5   2   1   0           4   \n",
       "Listeria_innocua_Cas9               0    0   67   41  23  13   5           8   \n",
       "N__meningitidis_Cas9                0    0    2    0   0   0   0           0   \n",
       "Neisseria_cinerea_Cas9              0    0    2    0   0   0   0           0   \n",
       "P__multocida_Cas9                   0    0    2    1   0   0   0           0   \n",
       "Parvibaculum_lavamentivorans_Cas9   0    0    0    0   0   0   0           0   \n",
       "S_CRISPR_1_thermophilus_Cas9        0    0    1    0   0   0   0           0   \n",
       "S_CRISPR_3_thermophilus_Cas9        2    0  170  101  59  30  13          86   \n",
       "S__mutans_Cas9                      2    0  131   91  59  33  15         124   \n",
       "S__pyogenes_Cas9                   10  NaN  130  104  79  54  31         408   \n",
       "Staphylococcus_aureus_Cas9          0    0    6    0   0   0   0           0   \n",
       "Streptococcus_pasteurianus_Cas9     0    0    2    0   0   0   0           1   \n",
       "T__denticola_Cas9                   0    0   27   12   2   1   0           3   \n",
       "\n",
       "                                  Total  \n",
       "C__jejuni_Cas9                        0  \n",
       "Campylobacter_lari_Cas9               0  \n",
       "Corynebacter_diphtheria_Cas9          0  \n",
       "F__novicida_Cas9                      0  \n",
       "Lactobacillus_buchneri_Cas9          23  \n",
       "Listeria_innocua_Cas9               149  \n",
       "N__meningitidis_Cas9                  2  \n",
       "Neisseria_cinerea_Cas9                2  \n",
       "P__multocida_Cas9                     3  \n",
       "Parvibaculum_lavamentivorans_Cas9     0  \n",
       "S_CRISPR_1_thermophilus_Cas9          1  \n",
       "S_CRISPR_3_thermophilus_Cas9        375  \n",
       "S__mutans_Cas9                      331  \n",
       "S__pyogenes_Cas9                    408  \n",
       "Staphylococcus_aureus_Cas9            6  \n",
       "Streptococcus_pasteurianus_Cas9       2  \n",
       "T__denticola_Cas9                    42  "
      ]
     },
     "execution_count": 630,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_dfs[74].loc['S__pyogenes_Cas9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/carlomazzaferro/Documents/Code/neoantigen/antigen_discovery/nepitope/pairwise_comp.py:193: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  df['Peptide'] = df['Peptide'].str.replace('X', '-')\n"
     ]
    }
   ],
   "source": [
    "pwcomp_1_swap = pairwise_comp.PairwiseComp(ls_mod, 5, mod_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_1_swap = pwcomp_1_swap.pipe_run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_1_swap.to_csv('/Users/carlomazzaferro/Desktop/Test_IEDB/OtherProtsPrashant/single_swap_pairwise_comp.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>Matches Loc</th>\n",
       "      <th>Num High AA</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C__jejuni_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Campylobacter_lari_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Corynebacter_diphtheria_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F__novicida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lactobacillus_buchneri_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[1266, 1267, 1268, 1269, 1270], [1266, 1267,...</td>\n",
       "      <td>5</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Listeria_innocua_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>66</td>\n",
       "      <td>40</td>\n",
       "      <td>23</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>118</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N__meningitidis_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neisseria_cinerea_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P__multocida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Parvibaculum_lavamentivorans_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__mutans_Cas9</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>119</td>\n",
       "      <td>81</td>\n",
       "      <td>49</td>\n",
       "      <td>24</td>\n",
       "      <td>10</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>202</td>\n",
       "      <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__pyogenes_Cas9</th>\n",
       "      <td>48</td>\n",
       "      <td>3</td>\n",
       "      <td>513</td>\n",
       "      <td>415</td>\n",
       "      <td>320</td>\n",
       "      <td>227</td>\n",
       "      <td>136</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>1659</td>\n",
       "      <td>1662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__thermophilus_CRISPR_3_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>160</td>\n",
       "      <td>93</td>\n",
       "      <td>52</td>\n",
       "      <td>23</td>\n",
       "      <td>8</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>210</td>\n",
       "      <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Staphylococcus_aureus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11], [650, ...</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus_pasteurianus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [1270, 1271, 1272, 1273, ...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T__denticola_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[164, 165, 166, 167, 168], [810, 811, 812, 8...</td>\n",
       "      <td>41</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   10 11    5    6    7    8    9  \\\n",
       "C__jejuni_Cas9                      0  0    0    0    0    0    0   \n",
       "Campylobacter_lari_Cas9             0  0    0    0    0    0    0   \n",
       "Corynebacter_diphtheria_Cas9        0  0    0    0    0    0    0   \n",
       "F__novicida_Cas9                    0  0    0    0    0    0    0   \n",
       "Lactobacillus_buchneri_Cas9         0  0   13    5    2    1    0   \n",
       "Listeria_innocua_Cas9               0  0   66   40   23   13    5   \n",
       "N__meningitidis_Cas9                0  0    2    0    0    0    0   \n",
       "Neisseria_cinerea_Cas9              0  0    2    0    0    0    0   \n",
       "P__multocida_Cas9                   0  0    2    1    0    0    0   \n",
       "Parvibaculum_lavamentivorans_Cas9   0  0    0    0    0    0    0   \n",
       "S__mutans_Cas9                      1  0  119   81   49   24   10   \n",
       "S__pyogenes_Cas9                   48  3  513  415  320  227  136   \n",
       "S__thermophilus_CRISPR_3_Cas9       0  0  160   93   52   23    8   \n",
       "Staphylococcus_aureus_Cas9          0  0    6    0    0    0    0   \n",
       "Streptococcus_pasteurianus_Cas9     0  0    2    0    0    0    0   \n",
       "T__denticola_Cas9                   0  0   27   12    2    1    0   \n",
       "\n",
       "                                                                         Matches Loc  \\\n",
       "C__jejuni_Cas9                                                                  [[]]   \n",
       "Campylobacter_lari_Cas9                                                         [[]]   \n",
       "Corynebacter_diphtheria_Cas9                                                    [[]]   \n",
       "F__novicida_Cas9                                                                [[]]   \n",
       "Lactobacillus_buchneri_Cas9        [[[1266, 1267, 1268, 1269, 1270], [1266, 1267,...   \n",
       "Listeria_innocua_Cas9              [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "N__meningitidis_Cas9                        [[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]   \n",
       "Neisseria_cinerea_Cas9                      [[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]   \n",
       "P__multocida_Cas9                                                               [[]]   \n",
       "Parvibaculum_lavamentivorans_Cas9                                               [[]]   \n",
       "S__mutans_Cas9                     [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "S__pyogenes_Cas9                   [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "S__thermophilus_CRISPR_3_Cas9      [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "Staphylococcus_aureus_Cas9         [[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11], [650, ...   \n",
       "Streptococcus_pasteurianus_Cas9    [[[7, 8, 9, 10, 11], [1270, 1271, 1272, 1273, ...   \n",
       "T__denticola_Cas9                  [[[164, 165, 166, 167, 168], [810, 811, 812, 8...   \n",
       "\n",
       "                                  Num High AA Total  \n",
       "C__jejuni_Cas9                              0     0  \n",
       "Campylobacter_lari_Cas9                     0     0  \n",
       "Corynebacter_diphtheria_Cas9                0     0  \n",
       "F__novicida_Cas9                            0     0  \n",
       "Lactobacillus_buchneri_Cas9                 5    21  \n",
       "Listeria_innocua_Cas9                     118   147  \n",
       "N__meningitidis_Cas9                        2     2  \n",
       "Neisseria_cinerea_Cas9                      2     2  \n",
       "P__multocida_Cas9                           0     3  \n",
       "Parvibaculum_lavamentivorans_Cas9           0     0  \n",
       "S__mutans_Cas9                            202   284  \n",
       "S__pyogenes_Cas9                         1659  1662  \n",
       "S__thermophilus_CRISPR_3_Cas9             210   336  \n",
       "Staphylococcus_aureus_Cas9                  6     6  \n",
       "Streptococcus_pasteurianus_Cas9             2     2  \n",
       "T__denticola_Cas9                          41    42  "
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_1_swap.loc['S__pyogenes_Cas9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>Matches Loc</th>\n",
       "      <th>Num High AA</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C__jejuni_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Campylobacter_lari_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Corynebacter_diphtheria_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F__novicida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lactobacillus_buchneri_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[1266, 1267, 1268, 1269, 1270], [1266, 1267,...</td>\n",
       "      <td>7</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Listeria_innocua_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>41</td>\n",
       "      <td>23</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>120</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N__meningitidis_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neisseria_cinerea_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P__multocida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Parvibaculum_lavamentivorans_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__mutans_Cas9</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>131</td>\n",
       "      <td>91</td>\n",
       "      <td>59</td>\n",
       "      <td>33</td>\n",
       "      <td>15</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>249</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__pyogenes_Cas9</th>\n",
       "      <td>52</td>\n",
       "      <td>3</td>\n",
       "      <td>542</td>\n",
       "      <td>441</td>\n",
       "      <td>343</td>\n",
       "      <td>245</td>\n",
       "      <td>147</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>1770</td>\n",
       "      <td>1773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__thermophilus_CRISPR_3_Cas9</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>171</td>\n",
       "      <td>101</td>\n",
       "      <td>59</td>\n",
       "      <td>30</td>\n",
       "      <td>13</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>250</td>\n",
       "      <td>376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Staphylococcus_aureus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11], [650, ...</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus_pasteurianus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [1270, 1271, 1272, 1273, ...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T__denticola_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[164, 165, 166, 167, 168], [810, 811, 812, 8...</td>\n",
       "      <td>41</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   10 11    5    6    7    8    9  \\\n",
       "C__jejuni_Cas9                      0  0    0    0    0    0    0   \n",
       "Campylobacter_lari_Cas9             0  0    0    0    0    0    0   \n",
       "Corynebacter_diphtheria_Cas9        0  0    0    0    0    0    0   \n",
       "F__novicida_Cas9                    0  0    0    0    0    0    0   \n",
       "Lactobacillus_buchneri_Cas9         0  0   15    5    2    1    0   \n",
       "Listeria_innocua_Cas9               0  0   67   41   23   13    5   \n",
       "N__meningitidis_Cas9                0  0    2    0    0    0    0   \n",
       "Neisseria_cinerea_Cas9              0  0    2    0    0    0    0   \n",
       "P__multocida_Cas9                   0  0    2    1    0    0    0   \n",
       "Parvibaculum_lavamentivorans_Cas9   0  0    0    0    0    0    0   \n",
       "S__mutans_Cas9                      2  0  131   91   59   33   15   \n",
       "S__pyogenes_Cas9                   52  3  542  441  343  245  147   \n",
       "S__thermophilus_CRISPR_3_Cas9       2  0  171  101   59   30   13   \n",
       "Staphylococcus_aureus_Cas9          0  0    6    0    0    0    0   \n",
       "Streptococcus_pasteurianus_Cas9     0  0    2    0    0    0    0   \n",
       "T__denticola_Cas9                   0  0   27   12    2    1    0   \n",
       "\n",
       "                                                                         Matches Loc  \\\n",
       "C__jejuni_Cas9                                                                  [[]]   \n",
       "Campylobacter_lari_Cas9                                                         [[]]   \n",
       "Corynebacter_diphtheria_Cas9                                                    [[]]   \n",
       "F__novicida_Cas9                                                                [[]]   \n",
       "Lactobacillus_buchneri_Cas9        [[[1266, 1267, 1268, 1269, 1270], [1266, 1267,...   \n",
       "Listeria_innocua_Cas9              [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "N__meningitidis_Cas9                        [[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]   \n",
       "Neisseria_cinerea_Cas9                      [[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]   \n",
       "P__multocida_Cas9                                                               [[]]   \n",
       "Parvibaculum_lavamentivorans_Cas9                                               [[]]   \n",
       "S__mutans_Cas9                     [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "S__pyogenes_Cas9                   [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "S__thermophilus_CRISPR_3_Cas9      [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "Staphylococcus_aureus_Cas9         [[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11], [650, ...   \n",
       "Streptococcus_pasteurianus_Cas9    [[[7, 8, 9, 10, 11], [1270, 1271, 1272, 1273, ...   \n",
       "T__denticola_Cas9                  [[[164, 165, 166, 167, 168], [810, 811, 812, 8...   \n",
       "\n",
       "                                  Num High AA Total  \n",
       "C__jejuni_Cas9                              0     0  \n",
       "Campylobacter_lari_Cas9                     0     0  \n",
       "Corynebacter_diphtheria_Cas9                0     0  \n",
       "F__novicida_Cas9                            0     0  \n",
       "Lactobacillus_buchneri_Cas9                 7    23  \n",
       "Listeria_innocua_Cas9                     120   149  \n",
       "N__meningitidis_Cas9                        2     2  \n",
       "Neisseria_cinerea_Cas9                      2     2  \n",
       "P__multocida_Cas9                           0     3  \n",
       "Parvibaculum_lavamentivorans_Cas9           0     0  \n",
       "S__mutans_Cas9                            249   331  \n",
       "S__pyogenes_Cas9                         1770  1773  \n",
       "S__thermophilus_CRISPR_3_Cas9             250   376  \n",
       "Staphylococcus_aureus_Cas9                  6     6  \n",
       "Streptococcus_pasteurianus_Cas9             2     2  \n",
       "T__denticola_Cas9                          41    42  "
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pairwise_comps.loc['S__pyogenes_Cas9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [py35]",
   "language": "python",
   "name": "Python [py35]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
