{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Conservation Analysis and Epitope Prediction\n",
    "\n",
    "\n",
    "#### Author: C. Mazzaferro, K. Fisch\n",
    "#### Email: cmazzafe@ucsd.edu\n",
    "#### Date: October 2016\n",
    " \n",
    "## Outline of Notebook\n",
    "<a id = \"toc\"></a>\n",
    "1. <a href = \"#background\">Background</a>\n",
    "2. <a href = \"#Cons\">High Affinity Binding Prediction </a>\n",
    "    * <a href = \"#Agg\">Data Aggregation</a>\n",
    "    * <a href = \"#Sim\">Similarty Score Calculation</a>\n",
    "    * <a href = \"#Plot\">Visualize</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Files of interest\n",
    "From a simple fasta file containing the proteins of interest, this workflow will create:\n",
    "\n",
    "1. Binding affinity prediction xls files for the given allele, nmer combinations. Default set to all allele supertypes and [8, 9, 10, 11] nmers\n",
    "2. Csv file containing all high binding affinity peptides. Threshold can be set to a customized level (either by setting it to Medium or setting it by affinity score. See [CBS's prediction services](http://www.cbs.dtu.dk/services/), in particular [netMHC](http://www.cbs.dtu.dk/services/NetMHC/) for more specific information.\n",
    "3. Fasta files containing swap data: for each high affinity binding peptide, find every peptide that results from swapping one single AA from it. These fasta files are then fed into netMHC again and their binding affinity determined again. \n",
    "4. Csv files for each fasta file with the swapped peptide binding prediction.\n",
    "\n",
    "Caveats: netMHC local install must be in the system's path. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/carlomazzaferro/anaconda/envs/py35/lib/python3.5/site-packages/matplotlib/__init__.py:1035: UserWarning: Duplicate key in file \"/Users/carlomazzaferro/.matplotlib/matplotlibrc\", line #2\n",
      "  (fname, cnt))\n"
     ]
    }
   ],
   "source": [
    "import importlib\n",
    "from nepitope import net_MHC_interface\n",
    "from nepitope import mhc_utils\n",
    "from nepitope import pep_utils\n",
    "import pandas\n",
    "\n",
    "project_location = '/Users/carlomazzaferro/Desktop/New_General/MICE/'\n",
    "fasta_file = '/Users/carlomazzaferro/Desktop/New_General/MICE/fasta_base_new_prots.fasta'\n",
    "net_mhc_path = '/Users/carlomazzaferro/Desktop/BINF_Tools/netMHC-4.0/netMHC'\n",
    "\n",
    "\n",
    "alleles = ['H-2-Ld', 'H-2-Db', 'H-2-Dd', 'H-2-Kb', 'H-2-Kd', 'H-2-Kk']\n",
    "nmers =[8, 9, 10, 11]\n",
    "\n",
    "net_mhc = net_MHC_interface.netMHCComand(net_mhc_path, fasta_file, nmers=nmers, alleles=alleles)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id = \"Cons\"></a>\n",
    "# MHC Binding Affinity Prediction \n",
    "### Ran on netMHCcons for n-mers of 8, 9, 10, 11 amino acids and for the 12 allele supertype families\n",
    "\n",
    "Run for each protein sequence retrieved from reference paper."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'netMHC commands written to run_netMHC.txt located at /Users/carlomazzaferro/Desktop/New_General/MICE/mhc_preds_fasta_base_new_prots'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "net_mhc.create_text_command()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predictions being saved to /Users/carlomazzaferro/Desktop/New_General/MICE/mhc_preds_fasta_base_new_prots\n"
     ]
    }
   ],
   "source": [
    "net_mhc.run_netMHC()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Load predictions \n",
    "import importlib\n",
    "importlib.reload(mhc_utils)\n",
    "pred_loc = '/Users/carlomazzaferro/Desktop/New_General/MICE/mhc_preds_fasta_2_prots/'\n",
    "agg = mhc_utils.FileConsolidation(pred_loc, fasta_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/carlomazzaferro/anaconda/envs/py35/lib/python3.5/site-packages/pandas/core/generic.py:2773: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self[name] = value\n"
     ]
    }
   ],
   "source": [
    "csv_dir = '/Users/carlomazzaferro/Desktop/New_General/MICE/high_affinity_peps_ALL.csv'\n",
    "high_aa_df = agg.get_all_high_affinity_from_batch(csv_out=True, csv_dir=csv_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "list_df = agg.return_df_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "lsss = agg.optimized_list_df_by_prot(list_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pandas.concat(lsss).to_csv('/Users/carlomazzaferro/Desktop/New_General/MICE/two_prots_all_preds.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from nepitope import mhc_utils, summary_data\n",
    "import glob\n",
    "nmers = [8, 9,10,11]\n",
    "\n",
    "#files saved from netMHCcons as split_*i*_prediction_*n*_mer for each split/nmer\n",
    "filepath = '/Users/carlomazzaferro/Desktop/new/swaps/mhc_preds_per_swap/'\n",
    "fasta_file = '/Users/carlomazzaferro/Desktop/Test_IEDB/newer_pipeline_files/fasta_from_study_dealigned.fasta'\n",
    "\n",
    "file_pattern = '_swap_*'\n",
    "\n",
    "\n",
    "files = glob.glob(filepath + file_pattern)\n",
    "original_peps_and_pos = []\n",
    "\n",
    "for file in files:\n",
    "    split_ = file.split('/')[-1]\n",
    "    split_2 = split_.split('_')\n",
    "    original_peps_and_pos.append([split_2[2], split_2[4]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from nepitope import summary_data\n",
    "importlib.reload(summary_data)\n",
    "smm = summary_data.SummaryData()\n",
    "ssmm = smm.summarize_all_data(lsss, fasta_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df = pandas.read_csv('/Users/carlomazzaferro/Desktop/New_General/MICE/high_affinity_peps.csv').sort_values(by='Pos')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_sp = df[df['ID'] == 'S__pyogenes_Cas9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_sp = df_sp.drop('Unnamed: 0', 1)\n",
    "df_sp.to_csv('/Users/carlomazzaferro/Desktop/New_General/MICE/SP_high_aff_peps.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "dat = list(zip(list(df_sp.Pos), list(df_sp.Peptide)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'I'"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[153]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein</th>\n",
       "      <th>Num High Affinity Peps</th>\n",
       "      <th>Num Med Affinity Peps</th>\n",
       "      <th>Num Low Affinity Peps</th>\n",
       "      <th>Num No Affinity Peps</th>\n",
       "      <th>Protein Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S__pyogenes_Cas9</td>\n",
       "      <td>40</td>\n",
       "      <td>174</td>\n",
       "      <td>673</td>\n",
       "      <td>31741</td>\n",
       "      <td>1368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Staphylococcus_aureus_Cas9</td>\n",
       "      <td>32</td>\n",
       "      <td>123</td>\n",
       "      <td>553</td>\n",
       "      <td>24360</td>\n",
       "      <td>1053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S_CRISPR_1_thermophilus_Cas9</td>\n",
       "      <td>25</td>\n",
       "      <td>100</td>\n",
       "      <td>539</td>\n",
       "      <td>26036</td>\n",
       "      <td>1121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>N__meningitidis_Cas9</td>\n",
       "      <td>20</td>\n",
       "      <td>89</td>\n",
       "      <td>424</td>\n",
       "      <td>25231</td>\n",
       "      <td>1082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Parvibaculum_lavamentivorans_Cas9</td>\n",
       "      <td>14</td>\n",
       "      <td>58</td>\n",
       "      <td>321</td>\n",
       "      <td>24291</td>\n",
       "      <td>1037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Corynebacter_diphtheria_Cas9</td>\n",
       "      <td>14</td>\n",
       "      <td>78</td>\n",
       "      <td>407</td>\n",
       "      <td>25313</td>\n",
       "      <td>1084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Streptococcus_pasteurianus_Cas9</td>\n",
       "      <td>34</td>\n",
       "      <td>101</td>\n",
       "      <td>469</td>\n",
       "      <td>26312</td>\n",
       "      <td>1130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Neisseria_cinerea_Cas9</td>\n",
       "      <td>22</td>\n",
       "      <td>86</td>\n",
       "      <td>442</td>\n",
       "      <td>25214</td>\n",
       "      <td>1082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Campylobacter_lari_Cas9</td>\n",
       "      <td>28</td>\n",
       "      <td>122</td>\n",
       "      <td>566</td>\n",
       "      <td>23152</td>\n",
       "      <td>1003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>T__denticola_Cas9</td>\n",
       "      <td>40</td>\n",
       "      <td>173</td>\n",
       "      <td>700</td>\n",
       "      <td>32363</td>\n",
       "      <td>1395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>S__mutans_Cas9</td>\n",
       "      <td>35</td>\n",
       "      <td>167</td>\n",
       "      <td>704</td>\n",
       "      <td>31170</td>\n",
       "      <td>1345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>S_CRISPR_3_thermophilus_Cas9</td>\n",
       "      <td>45</td>\n",
       "      <td>156</td>\n",
       "      <td>715</td>\n",
       "      <td>32192</td>\n",
       "      <td>1388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>C__jejuni_Cas9</td>\n",
       "      <td>26</td>\n",
       "      <td>99</td>\n",
       "      <td>527</td>\n",
       "      <td>22760</td>\n",
       "      <td>984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>P__multocida_Cas9</td>\n",
       "      <td>32</td>\n",
       "      <td>96</td>\n",
       "      <td>517</td>\n",
       "      <td>24494</td>\n",
       "      <td>1056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>F__novicida_Cas9</td>\n",
       "      <td>48</td>\n",
       "      <td>195</td>\n",
       "      <td>803</td>\n",
       "      <td>37845</td>\n",
       "      <td>1629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Lactobacillus_buchneri_Cas9</td>\n",
       "      <td>22</td>\n",
       "      <td>144</td>\n",
       "      <td>701</td>\n",
       "      <td>31833</td>\n",
       "      <td>1371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Listeria_innocua_Cas9</td>\n",
       "      <td>39</td>\n",
       "      <td>152</td>\n",
       "      <td>697</td>\n",
       "      <td>30924</td>\n",
       "      <td>1334</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Protein Num High Affinity Peps  \\\n",
       "0                    S__pyogenes_Cas9                     40   \n",
       "1          Staphylococcus_aureus_Cas9                     32   \n",
       "2        S_CRISPR_1_thermophilus_Cas9                     25   \n",
       "3                N__meningitidis_Cas9                     20   \n",
       "4   Parvibaculum_lavamentivorans_Cas9                     14   \n",
       "5        Corynebacter_diphtheria_Cas9                     14   \n",
       "6     Streptococcus_pasteurianus_Cas9                     34   \n",
       "7              Neisseria_cinerea_Cas9                     22   \n",
       "8             Campylobacter_lari_Cas9                     28   \n",
       "9                   T__denticola_Cas9                     40   \n",
       "10                     S__mutans_Cas9                     35   \n",
       "11       S_CRISPR_3_thermophilus_Cas9                     45   \n",
       "12                     C__jejuni_Cas9                     26   \n",
       "13                  P__multocida_Cas9                     32   \n",
       "14                   F__novicida_Cas9                     48   \n",
       "15        Lactobacillus_buchneri_Cas9                     22   \n",
       "16              Listeria_innocua_Cas9                     39   \n",
       "\n",
       "   Num Med Affinity Peps Num Low Affinity Peps Num No Affinity Peps  \\\n",
       "0                    174                   673                31741   \n",
       "1                    123                   553                24360   \n",
       "2                    100                   539                26036   \n",
       "3                     89                   424                25231   \n",
       "4                     58                   321                24291   \n",
       "5                     78                   407                25313   \n",
       "6                    101                   469                26312   \n",
       "7                     86                   442                25214   \n",
       "8                    122                   566                23152   \n",
       "9                    173                   700                32363   \n",
       "10                   167                   704                31170   \n",
       "11                   156                   715                32192   \n",
       "12                    99                   527                22760   \n",
       "13                    96                   517                24494   \n",
       "14                   195                   803                37845   \n",
       "15                   144                   701                31833   \n",
       "16                   152                   697                30924   \n",
       "\n",
       "   Protein Length  \n",
       "0            1368  \n",
       "1            1053  \n",
       "2            1121  \n",
       "3            1082  \n",
       "4            1037  \n",
       "5            1084  \n",
       "6            1130  \n",
       "7            1082  \n",
       "8            1003  \n",
       "9            1395  \n",
       "10           1345  \n",
       "11           1388  \n",
       "12            984  \n",
       "13           1056  \n",
       "14           1629  \n",
       "15           1371  \n",
       "16           1334  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pandas.DataFrame(ssmm.data_list, index=ssmm.indexes).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'high_aa_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-b6c06b421a81>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mhigh_aa_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'high_aa_df' is not defined"
     ]
    }
   ],
   "source": [
    "high_aa_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Swap Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "csv_dir = '/Users/carlomazzaferro/Desktop/New_General/MICE/high_affinity_peps.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "high_aa_df = pandas.read_csv(csv_dir) \n",
    "high_aa_df = high_aa_df[high_aa_df['ID'] == 'Staphylococcus_aureus_Cas9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from nepitope import peptide_utilities\n",
    "importlib.reload(peptide_utilities)\n",
    "fasta_files_dir = '/Users/carlomazzaferro/Desktop/swaps_2/'\n",
    "\n",
    "swap = peptide_utilities.Swaps(high_aa_df, fasta_files_dir, net_mhc_path, proteins=['Staphylococcus_aureus_Cas9'] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_ALIIANADFI_Pos_704_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Db_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_ITYREYLENM_Pos_998_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kb_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_SEEEFSAALL_Pos_100_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_EEFSAALLHL_Pos_102_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_LEYYEKFQII_Pos_274_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_IETEQEYKEI_Pos_743_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_TEQEYKEIFI_Pos_745_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_YEEAKKLKKI_Pos_946_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_10\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_IAIFNRLKL_Pos_414_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Db_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_LIIANADFI_Pos_705_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Db_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_VTVKNLDVI_Pos_925_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Db_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_ASFYNNDLI_Pos_963_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Db_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_RSYFRVNNL_Pos_660_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kb_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_SYETFKKHI_Pos_599_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kd_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_YYEETGNYL_Pos_856_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kd_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_SEEEFSAAL_Pos_100_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_EEEFSAALL_Pos_101_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_YEMLMGHCT_Pos_229_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_LEYYEKFQI_Pos_274_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_FEEKQAESM_Pos_732_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_EEAKKLKKI_Pos_947_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_9\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_LSEEEFSAALL_Pos_99_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_11\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_YEMLMGHCTYF_Pos_229_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_11\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_ETEQEYKEIFI_Pos_744_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_11\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_EEFSAALL_Pos_102_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_8\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_YEMLMGHC_Pos_229_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_8\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_KENAKYLI_Pos_513_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_8\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_YETFKKHI_Pos_600_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_8\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_TEQEYKEI_Pos_745_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_8\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_QEYKEIFI_Pos_747_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_8\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_NDKLKKLI_Pos_812_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_8\n",
      "Predictions being saved to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/mhc_preds__swap_YEETGNYL_Pos_857_ID_Staphylococcus-aureus-Cas9_Allele_H-2-Kk_nmer_8\n",
      "Swap predictions regrouped to /Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/preds_per_swap\n"
     ]
    }
   ],
   "source": [
    "cmd = swap.find_swaps_write_to_fasta()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from nepitope import mhc_utils, summary_data\n",
    "import glob\n",
    "nmers = [8, 9,10,11]\n",
    "\n",
    "#files saved from netMHCcons as split_*i*_prediction_*n*_mer for each split/nmer\n",
    "filepath = '/Users/carlomazzaferro/Desktop/New_General/MICE/mhc_preds_fasta_2_prots/'\n",
    "fasta_file = '/Users/carlomazzaferro/Desktop/New_General/MICE/fasta_2_prots.fasta'\n",
    "\n",
    "file_pattern = '_swap_*'\n",
    "\n",
    "aggregate_all = mhc_utils.FileConsolidation(filepath, fasta_file)\n",
    "\n",
    "files = glob.glob(filepath + file_pattern)\n",
    "original_peps_and_pos = []\n",
    "\n",
    "for file in files:\n",
    "    split_ = file.split('/')[-1]\n",
    "    split_2 = split_.split('_')\n",
    "    original_peps_and_pos.append([split_2[2], split_2[4]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from nepitope import mhc_utils\n",
    "import pandas\n",
    "import glob\n",
    "import importlib\n",
    "importlib.reload(mhc_utils)\n",
    "\n",
    "#files saved from netMHC in  multiple different files, one per each query (nmer, allele)\n",
    "filepath = '/Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/preds_per_swap/'\n",
    "fasta_file = '/Users/carlomazzaferro/Desktop/New_General/MICE/fasta_base_new_prots.fasta'\n",
    "aggregate_all = mhc_utils.FileConsolidation(filepath, fasta_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "files = glob.glob(filepath + '*')\n",
    "original_peps_and_pos = []\n",
    "\n",
    "for file in files:\n",
    "    split_ = file.split('/')[-1]\n",
    "    split_2 = split_.split('_')\n",
    "    original_peps_and_pos.append([split_2[2], split_2[4]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#returns a dataframe per file used (just one in this case)\n",
    "df_list = aggregate_all.return_df_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from nepitope import mhc_utils, summary_data\n",
    "\n",
    "summ_data = summary_data.SummaryData()\n",
    "summ_data_analysis = summ_data.summarize_data_for_each_mhc_pred(df_list, original_peps_and_pos, high_aa_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "csv_path = '/Users/carlomazzaferro/Desktop/swaps_2/Staphylococcus_aureus_Cas9/summary_results_per_prediction.csv'\n",
    "summ_data_analysis.summary_df.sort_values(by='original pos').to_csv(path_or_buf=csv_path, sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ss = summ_data_analysis.summary_df.sort_values(by='original pos')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DFLIST FROM ALL PREDS "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_list = agg.return_df_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "lsss_1 = agg.optimized_list_df_by_prot(df_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/carlomazzaferro/Documents/Code/neoantigen/antigen_discovery/nepitope/pairwise_comp.py:194: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  df['Peptide'] = df['Peptide'].str.replace('X', '-')\n"
     ]
    }
   ],
   "source": [
    "from nepitope import pairwise_comp\n",
    "importlib.reload(pairwise_comp)\n",
    "\n",
    "fasta_file = '/Users/carlomazzaferro/Desktop/New_General/MICE/fasta_base_new_prots.fasta'\n",
    "\n",
    "\n",
    "pwcomp = pairwise_comp.PairwiseComp(lsss_1, 5, fasta_file) \n",
    "df_pairwise_comps  = pwcomp.pipe_run() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sa_df = pwcomp.filt_dfs[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sa_peps = sa_df.Peptide.values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def _get_possible_peptides(stri):\n",
    "    # TODO: Clean this up asap. Too hacky, not very readable\n",
    "\n",
    "    nmer_list = list(range(5, len(stri) + 1))\n",
    "    length_pep = len(stri)\n",
    "    num_possible_sub_peps = [length_pep - nmer + 1 for nmer in nmer_list if length_pep - nmer + 1 > 0]\n",
    "\n",
    "    subs_all = []\n",
    "    for i in num_possible_sub_peps:\n",
    "        for nmer, j in zip(nmer_list, range(i)):\n",
    "            subs = [stri[j:j + nmer] for i, nmer in enumerate(nmer_list)]\n",
    "            subs_all.extend(subs)\n",
    "\n",
    "    subs_all = list(set(subs_all))\n",
    "    lst = [[w for w in subs_all if len(w) == num] for num in set(len(i) for i in subs_all)]\n",
    "\n",
    "    return lst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "all_5_nmers = {}\n",
    "for pep in sa_peps:\n",
    "    all_5_nmers[pep] = _get_possible_peptides(pep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "finder = []\n",
    "for i in list(all_5_nmers.values()):\n",
    "    for j in i:\n",
    "        for k in j:\n",
    "            if prot_sp.find(k) != -1:\n",
    "                finder.append(prot_sp.find(k))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[346, 346, 346]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "finder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "152       [[152, 153, 154, 155, 156, 157, 158, 159, 160]]\n",
       "153       [[153, 154, 155, 156, 157, 158, 159, 160, 161]]\n",
       "367       [[367, 368, 369, 370, 371, 372, 373, 374, 375]]\n",
       "368            [[368, 369, 370, 371, 372, 373, 374, 375]]\n",
       "368     [[368, 369, 370, 371, 372, 373, 374, 375, 376,...\n",
       "369     [[369, 370, 371, 372, 373, 374, 375, 376, 377,...\n",
       "369     [[369, 370, 371, 372, 373, 374, 375, 376, 377,...\n",
       "385            [[385, 386, 387, 388, 389, 390, 391, 392]]\n",
       "439       [[439, 440, 441, 442, 443, 444, 445, 446, 447]]\n",
       "515       [[515, 516, 517, 518, 519, 520, 521, 522, 523]]\n",
       "520     [[520, 521, 522, 523, 524, 525, 526, 527, 528,...\n",
       "521       [[521, 522, 523, 524, 525, 526, 527, 528, 529]]\n",
       "564            [[564, 565, 566, 567, 568, 569, 570, 571]]\n",
       "687       [[687, 688, 689, 690, 691, 692, 693, 694, 695]]\n",
       "688            [[688, 689, 690, 691, 692, 693, 694, 695]]\n",
       "807       [[807, 808, 809, 810, 811, 812, 813, 814, 815]]\n",
       "891       [[891, 892, 893, 894, 895, 896, 897, 898, 899]]\n",
       "906     [[906, 907, 908, 909, 910, 911, 912, 913, 914,...\n",
       "907     [[907, 908, 909, 910, 911, 912, 913, 914, 915,...\n",
       "907     [[907, 908, 909, 910, 911, 912, 913, 914, 915,...\n",
       "908     [[908, 909, 910, 911, 912, 913, 914, 915, 916,...\n",
       "908       [[908, 909, 910, 911, 912, 913, 914, 915, 916]]\n",
       "908     [[908, 909, 910, 911, 912, 913, 914, 915, 916,...\n",
       "909            [[909, 910, 911, 912, 913, 914, 915, 916]]\n",
       "1031    [[1031, 1032, 1033, 1034, 1035, 1036, 1037, 10...\n",
       "1032    [[1032, 1033, 1034, 1035, 1036, 1037, 1038, 10...\n",
       "1032    [[1032, 1033, 1034, 1035, 1036, 1037, 1038, 10...\n",
       "1033    [[1033, 1034, 1035, 1036, 1037, 1038, 1039, 10...\n",
       "1046    [[1046, 1047, 1048, 1049, 1050, 1051, 1052, 10...\n",
       "1047    [[1047, 1048, 1049, 1050, 1051, 1052, 1053, 10...\n",
       "1054    [[1054, 1055, 1056, 1057, 1058, 1059, 1060, 10...\n",
       "1160    [[1160, 1161, 1162, 1163, 1164, 1165, 1166, 11...\n",
       "1197    [[1197, 1198, 1199, 1200, 1201, 1202, 1203, 12...\n",
       "1223    [[1223, 1224, 1225, 1226, 1227, 1228, 1229, 12...\n",
       "1251    [[1251, 1252, 1253, 1254, 1255, 1256, 1257, 12...\n",
       "1273    [[1273, 1274, 1275, 1276, 1277, 1278, 1279, 12...\n",
       "1337    [[1337, 1338, 1339, 1340, 1341, 1342, 1343, 13...\n",
       "1338    [[1338, 1339, 1340, 1341, 1342, 1343, 1344, 13...\n",
       "1339    [[1339, 1340, 1341, 1342, 1343, 1344, 1345, 13...\n",
       "1340    [[1340, 1341, 1342, 1343, 1344, 1345, 1346, 13...\n",
       "Name: Range, dtype: object"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sp_df.Range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sp_df = pwcomp.filt_dfs[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "prot_sp = pwcomp.peps_prots_ranges[0][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>Matches Loc</th>\n",
       "      <th>Num High AA</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C__jejuni_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Campylobacter_lari_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Corynebacter_diphtheria_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F__novicida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lactobacillus_buchneri_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Listeria_innocua_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>14</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[155, 156, 157, 158, 159], [154, 155, 156, 1...</td>\n",
       "      <td>23</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N__meningitidis_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neisseria_cinerea_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P__multocida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Parvibaculum_lavamentivorans_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_CRISPR_1_thermophilus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_CRISPR_3_thermophilus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>31</td>\n",
       "      <td>16</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>[[[156, 157, 158, 159, 160], [155, 156, 157, 1...</td>\n",
       "      <td>97</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__mutans_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>13</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>[[[155, 156, 157, 158, 159], [154, 155, 156, 1...</td>\n",
       "      <td>28</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__pyogenes_Cas9</th>\n",
       "      <td>26</td>\n",
       "      <td>9</td>\n",
       "      <td>217</td>\n",
       "      <td>177</td>\n",
       "      <td>137</td>\n",
       "      <td>97</td>\n",
       "      <td>57</td>\n",
       "      <td>[[[152, 153, 154, 155, 156, 157, 158, 159], [1...</td>\n",
       "      <td>720</td>\n",
       "      <td>720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Staphylococcus_aureus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus_pasteurianus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T__denticola_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[810, 811, 812, 813, 814], [809, 810, 811, 8...</td>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   10 11    5    6    7   8   9  \\\n",
       "C__jejuni_Cas9                      0  0    0    0    0   0   0   \n",
       "Campylobacter_lari_Cas9             0  0    3    0    0   0   0   \n",
       "Corynebacter_diphtheria_Cas9        0  0    0    0    0   0   0   \n",
       "F__novicida_Cas9                    0  0    0    0    0   0   0   \n",
       "Lactobacillus_buchneri_Cas9         0  0    1    0    0   0   0   \n",
       "Listeria_innocua_Cas9               0  0   26   14    4   1   0   \n",
       "N__meningitidis_Cas9                0  0    0    0    0   0   0   \n",
       "Neisseria_cinerea_Cas9              0  0    3    0    0   0   0   \n",
       "P__multocida_Cas9                   0  0    0    0    0   0   0   \n",
       "Parvibaculum_lavamentivorans_Cas9   0  0    0    0    0   0   0   \n",
       "S_CRISPR_1_thermophilus_Cas9        0  0    0    0    0   0   0   \n",
       "S_CRISPR_3_thermophilus_Cas9        0  0   49   31   16   6   2   \n",
       "S__mutans_Cas9                      0  0   30   13    4   2   1   \n",
       "S__pyogenes_Cas9                   26  9  217  177  137  97  57   \n",
       "Staphylococcus_aureus_Cas9          0  0    3    0    0   0   0   \n",
       "Streptococcus_pasteurianus_Cas9     0  0    4    0    0   0   0   \n",
       "T__denticola_Cas9                   0  0   12    1    0   0   0   \n",
       "\n",
       "                                                                         Matches Loc  \\\n",
       "C__jejuni_Cas9                                                                  [[]]   \n",
       "Campylobacter_lari_Cas9                                                         [[]]   \n",
       "Corynebacter_diphtheria_Cas9                                                    [[]]   \n",
       "F__novicida_Cas9                                                                [[]]   \n",
       "Lactobacillus_buchneri_Cas9                                                     [[]]   \n",
       "Listeria_innocua_Cas9              [[[155, 156, 157, 158, 159], [154, 155, 156, 1...   \n",
       "N__meningitidis_Cas9                                                            [[]]   \n",
       "Neisseria_cinerea_Cas9                                                          [[]]   \n",
       "P__multocida_Cas9                                                               [[]]   \n",
       "Parvibaculum_lavamentivorans_Cas9                                               [[]]   \n",
       "S_CRISPR_1_thermophilus_Cas9                                                    [[]]   \n",
       "S_CRISPR_3_thermophilus_Cas9       [[[156, 157, 158, 159, 160], [155, 156, 157, 1...   \n",
       "S__mutans_Cas9                     [[[155, 156, 157, 158, 159], [154, 155, 156, 1...   \n",
       "S__pyogenes_Cas9                   [[[152, 153, 154, 155, 156, 157, 158, 159], [1...   \n",
       "Staphylococcus_aureus_Cas9                                                      [[]]   \n",
       "Streptococcus_pasteurianus_Cas9                                                 [[]]   \n",
       "T__denticola_Cas9                  [[[810, 811, 812, 813, 814], [809, 810, 811, 8...   \n",
       "\n",
       "                                  Num High AA Total  \n",
       "C__jejuni_Cas9                              0     0  \n",
       "Campylobacter_lari_Cas9                     0     3  \n",
       "Corynebacter_diphtheria_Cas9                0     0  \n",
       "F__novicida_Cas9                            0     0  \n",
       "Lactobacillus_buchneri_Cas9                 0     1  \n",
       "Listeria_innocua_Cas9                      23    45  \n",
       "N__meningitidis_Cas9                        0     0  \n",
       "Neisseria_cinerea_Cas9                      0     3  \n",
       "P__multocida_Cas9                           0     0  \n",
       "Parvibaculum_lavamentivorans_Cas9           0     0  \n",
       "S_CRISPR_1_thermophilus_Cas9                0     0  \n",
       "S_CRISPR_3_thermophilus_Cas9               97   104  \n",
       "S__mutans_Cas9                             28    50  \n",
       "S__pyogenes_Cas9                          720   720  \n",
       "Staphylococcus_aureus_Cas9                  0     3  \n",
       "Streptococcus_pasteurianus_Cas9             0     4  \n",
       "T__denticola_Cas9                          10    13  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref = 'S__pyogenes_Cas9'\n",
    "df_pairwise_comps.loc[ref]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_pairwise_comps.to_csv('/Users/carlomazzaferro/Desktop/New_General/MICE/df_pairwise_comps.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "swaps_df_dir = '/Users/carlomazzaferro/Desktop/Test_IEDB/OtherProtsPrashant/swap_results/S__pyogenes_Cas9/summary_results_per_prediction.csv'\n",
    "swaps_df = pandas.read_csv(swaps_df_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/carlomazzaferro/Documents/Code/neoantigen/antigen_discovery/nepitope/merge_workflows.py:120: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  df = pandas.DataFrame(maximal_list)\n"
     ]
    }
   ],
   "source": [
    "from importlib import reload\n",
    "reload(merge_workflows)\n",
    "from nepitope import merge_workflows\n",
    "merged = merge_workflows.MergeSwapsAndComp(df_pairwise_comps, swaps_df, 40)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 577,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "130 matches occurring at position 657\n",
      "129 matches occurring at position 447\n",
      "126 matches occurring at position 656\n",
      "124 matches occurring at position 446\n",
      "122 matches occurring at position 658\n",
      "119 matches occurring at position 448\n",
      "117 matches occurring at position 813\n",
      "115 matches occurring at position 814\n",
      "110 matches occurring at position 659\n",
      "109 matches occurring at position 445\n",
      "109 matches occurring at position 812\n",
      "106 matches occurring at position 514\n",
      "106 matches occurring at position 655\n",
      "103 matches occurring at position 513\n",
      "100 matches occurring at position 449\n",
      "98 matches occurring at position 512\n",
      "98 matches occurring at position 515\n",
      "98 matches occurring at position 985\n",
      "98 matches occurring at position 987\n",
      "96 matches occurring at position 986\n"
     ]
    }
   ],
   "source": [
    "merged.top_loc_matches[0:20]\n",
    "for i in merged.top_loc_matches[0:20]:\n",
    "    print('%i matches occurring at position %i' %(i[1], i[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/carlomazzaferro/Documents/Code/neoantigen/antigen_discovery/nepitope/merge_workflows.py:96: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "df_top = merged.top_swap_df\n",
    "df_ordered_filtered = merged.final_sort_and_filtering(df_top)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 642,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'nmer', 'allele', 'num high affinity peps',\n",
       "       'num med affinity peps', 'num low affinity peps',\n",
       "       'num no affinity peps', 'original peptide', 'original pos',\n",
       "       'top scoring peptides', 'Range'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 642,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ordered_filtered.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 610,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "reload(modify_data)\n",
    "from nepitope import modify_data\n",
    "mod_fasta = '/Users/carlomazzaferro/Desktop/PROJECT_LOC_TEST/mod_fastas/mod_all_fastas/'\n",
    "orig_fasta = '/Users/carlomazzaferro/Desktop/PROJECT_LOC_TEST/prots_in_fasta.fasta'\n",
    "\n",
    "mod_data = modify_data.ModifyData(swaps_df, fasta_file, mod_fasta, lsss_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 611,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "True\n",
      "RRYTRRKNR\n",
      "True\n",
      "KAILSARLSK\n",
      "True\n",
      "KLNREDLLRK\n",
      "True\n",
      "WRQLLNAKLI\n",
      "True\n",
      "SRFAWMTRK\n",
      "True\n",
      "LIHDDSLTF\n",
      "True\n",
      "IYLALAHMI\n",
      "True\n",
      "YHHAHDAYL\n",
      "True\n",
      "SLLYEYFTVY\n",
      "True\n",
      "ILLSDILRV\n",
      "True\n",
      "MLASAGELQK\n",
      "True\n",
      "KQSGKTILDF\n",
      "True\n",
      "RRLENLIAQL\n",
      "True\n",
      "KTYAHLFDDK\n",
      "True\n",
      "AHMIKFRGHF\n",
      "True\n",
      "LPKYSLFEL\n",
      "True\n",
      "SLHEHIANL\n",
      "True\n",
      "SPAIKKGIL\n",
      "True\n",
      "RLIYLALAHM\n",
      "True\n",
      "HYLDEIIEQI\n",
      "True\n",
      "RRKNRICYL\n",
      "True\n",
      "REQAENIIHL\n",
      "True\n",
      "KVLPKHSLLY\n",
      "True\n",
      "MIKFRGHFL\n",
      "True\n",
      "LLYEYFTVY\n",
      "True\n",
      "ALSLGLTPNF\n",
      "True\n",
      "NGLFGNLIAL\n",
      "True\n",
      "RRYTGWGRLS\n",
      "True\n",
      "LLYEYFTV\n",
      "True\n",
      "AHDAYLNAV\n",
      "True\n",
      "YLQNGRDMYV\n",
      "True\n",
      "TIMERSSFEK\n",
      "True\n",
      "RRYTRRKNRI\n",
      "True\n",
      "ILTFRIPYYV\n",
      "True\n",
      "RRRYTGWGR\n",
      "True\n",
      "SEFSKRVIL\n",
      "True\n",
      "GLDIGTNSV\n",
      "True\n",
      "MKNYWRQLL\n",
      "True\n",
      "RRRYTGWGRL\n",
      "True\n",
      "VEDRFNASL\n",
      "True\n",
      "FVYGDYKVY\n",
      "True\n",
      "SPTVAYSVL\n",
      "True\n",
      "IPHQIHLGEL\n",
      "True\n",
      "AYHEKYPTI\n",
      "True\n",
      "FANRNFMQL\n",
      "True\n",
      "KLPKYSLFEL\n",
      "True\n",
      "LTFRIPYYV\n",
      "True\n",
      "AENIIHLFTL\n",
      "True\n",
      "AILSARLSK\n",
      "True\n",
      "HMIKFRGHF\n",
      "True\n",
      "NSRFAWMTRK\n",
      "True\n",
      "EIIEQISEF\n",
      "True\n",
      "NLIALSLGL\n",
      "True\n",
      "FLYLASHYEK\n",
      "True\n",
      "AERGGLSEL\n",
      "True\n",
      "GVEDRFNASL\n",
      "True\n",
      "DEIIEQISEF\n",
      "True\n",
      "GELQKGNEL\n",
      "True\n",
      "EEVVKKMKNY\n",
      "True\n",
      "YLDEIIEQI\n",
      "True\n",
      "GLFGNLIAL\n",
      "True\n",
      "YEYFTVYNEL\n",
      "True\n",
      "VPSKKFKVL\n",
      "True\n",
      "EFVYGDYKVY\n",
      "True\n",
      "RRLENLIAQLP\n",
      "True\n",
      "RRQEDFYPFL\n",
      "True\n",
      "QLKRRRYTG\n",
      "True\n",
      "KSVKELLGI\n",
      "True\n",
      "SRFAWMTRKS\n",
      "True\n",
      "LQNEKLYLYY\n",
      "True\n",
      "LESEFVYGDY\n",
      "True\n",
      "RRYTGWGRL\n",
      "True\n",
      "MIKFRGHF\n",
      "True\n",
      "KTNRKVTVK\n",
      "True\n",
      "FYSNIMNFF\n",
      "True\n",
      "GETAEATRL\n",
      "True\n",
      "KVYDVRKMIAK\n",
      "True\n",
      "KRMLASAGEL\n",
      "True\n",
      "VSDFRKDFQFY\n",
      "True\n",
      "EIIEQISEFS\n",
      "True\n",
      "HSLLYEYFTV\n",
      "True\n",
      "YLASHYEKL\n",
      "True\n",
      "RRYTGWGR\n",
      "True\n",
      "SLLYEYFTV\n",
      "True\n",
      "TNLGAPAAFK\n",
      "True\n",
      "QLIHDDSLTF\n",
      "True\n",
      "KSEETITPW\n",
      "True\n",
      "NLDKVLSAY\n",
      "True\n",
      "KATAKYFFY\n",
      "True\n",
      "ESEFVYGDY\n",
      "True\n",
      "EVVKKMKNY\n",
      "True\n",
      "YLNAVVGTAL\n",
      "True\n",
      "SLGTYHDLLK\n",
      "True\n",
      "LIHQSITGLY\n",
      "True\n",
      "HMIKFRGHFL\n",
      "True\n",
      "KLFIQLVQTY\n",
      "True\n",
      "AILLSDILRV\n"
     ]
    }
   ],
   "source": [
    "######## TO INCLUDE IN TEST SUITE ########\n",
    "\n",
    "from nepitope import pep_utils\n",
    "j,k = pep_utils.create_separate_lists(fasta_file)\n",
    "for i in sgl:\n",
    "    print(i[1] in k[0])\n",
    "\n",
    "for i in mod_data.swaps_dic.keys():\n",
    "    print(i in mod_data.fasta_dic[0][1])\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 612,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mod_data.get_modified_fastas_single_swap()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 615,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exchange_pairs = []\n",
    "for i,j in mod_data.swaps_dic.items():\n",
    "    exchange_pairs.append([i,j])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 603,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "def get_all_exchange_pairs(swaps_df):\n",
    "    cols = ['original peptide', 'top scoring peptides']\n",
    "    sliced = swaps_df.ix[:, cols].values.tolist()\n",
    "    orig = [i[0] for i in sliced]\n",
    "    swaps = [list(filter(None, re.sub('[^A-Za-z0-9]+', ',', i[1]).split(','))) for i in sliced]\n",
    "    hashmap = [[orig[i], swaps[i]] for i in range(len(orig))]\n",
    "    swaps_single = []\n",
    "    for idx, hsh in enumerate(hashmap):\n",
    "        swaps_single.append([hsh[0], random.choice(hsh[1])])\n",
    "    return swaps_single\n",
    "\n",
    "sgl = get_all_exchange_pairs(swaps_df)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 616,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/carlomazzaferro/Documents/Code/neoantigen/antigen_discovery/nepitope/pairwise_comp.py:194: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  df['Peptide'] = df['Peptide'].str.replace('X', '-')\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "fastas = glob.glob(mod_fasta + '*')\n",
    "csv_path = '/Users/carlomazzaferro/Desktop/PROJECT_LOC_TEST/mod_csvs/mod_csvs_all_peps/'\n",
    "\n",
    "list_dfs = []\n",
    "for pair in exchange_pairs:\n",
    "    for fasta in fastas:\n",
    "        if os.path.splitext(os.path.basename(fasta))[0].split('-')[0] == pair[0]:\n",
    "            my_fasta = fasta\n",
    "            ls_mod = mod_data.get_singly_modified_df_list(pair[0], pair[1])\n",
    "            pwcomp_pair_swap = pairwise_comp.PairwiseComp(ls_mod, 5, my_fasta)\n",
    "            df_swap = pwcomp_pair_swap.pipe_run()\n",
    "            df_swap = df_swap.drop('Matches Loc', 1)\n",
    "            df_swap.to_csv(csv_path + pair[0] + '-' + pair[1] + '.csv')\n",
    "            list_dfs.append(df_swap)\n",
    "            \n",
    "            \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 629,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>Num High AA</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C__jejuni_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Campylobacter_lari_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Corynebacter_diphtheria_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F__novicida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lactobacillus_buchneri_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Listeria_innocua_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>41</td>\n",
       "      <td>23</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N__meningitidis_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neisseria_cinerea_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P__multocida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Parvibaculum_lavamentivorans_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_CRISPR_1_thermophilus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_CRISPR_3_thermophilus_Cas9</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>169</td>\n",
       "      <td>101</td>\n",
       "      <td>59</td>\n",
       "      <td>30</td>\n",
       "      <td>13</td>\n",
       "      <td>70</td>\n",
       "      <td>374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__mutans_Cas9</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>128</td>\n",
       "      <td>88</td>\n",
       "      <td>56</td>\n",
       "      <td>30</td>\n",
       "      <td>13</td>\n",
       "      <td>87</td>\n",
       "      <td>316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__pyogenes_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27</td>\n",
       "      <td>20</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Staphylococcus_aureus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus_pasteurianus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T__denticola_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  10   11    5    6   7   8   9 Num High AA  \\\n",
       "C__jejuni_Cas9                     0    0    0    0   0   0   0           0   \n",
       "Campylobacter_lari_Cas9            0    0    0    0   0   0   0           0   \n",
       "Corynebacter_diphtheria_Cas9       0    0    0    0   0   0   0           0   \n",
       "F__novicida_Cas9                   0    0    0    0   0   0   0           0   \n",
       "Lactobacillus_buchneri_Cas9        0    0   15    5   2   1   0           2   \n",
       "Listeria_innocua_Cas9              0    0   67   41  23  13   5           7   \n",
       "N__meningitidis_Cas9               0    0    2    0   0   0   0           0   \n",
       "Neisseria_cinerea_Cas9             0    0    2    0   0   0   0           0   \n",
       "P__multocida_Cas9                  0    0    2    1   0   0   0           0   \n",
       "Parvibaculum_lavamentivorans_Cas9  0    0    0    0   0   0   0           0   \n",
       "S_CRISPR_1_thermophilus_Cas9       0    0    1    0   0   0   0           0   \n",
       "S_CRISPR_3_thermophilus_Cas9       2    0  169  101  59  30  13          70   \n",
       "S__mutans_Cas9                     1    0  128   88  56  30  13          87   \n",
       "S__pyogenes_Cas9                   0  NaN   27   20  13   6   2          68   \n",
       "Staphylococcus_aureus_Cas9         0    0    6    0   0   0   0           0   \n",
       "Streptococcus_pasteurianus_Cas9    0    0    2    0   0   0   0           0   \n",
       "T__denticola_Cas9                  0    0   27   12   2   1   0           1   \n",
       "\n",
       "                                  Total  \n",
       "C__jejuni_Cas9                        0  \n",
       "Campylobacter_lari_Cas9               0  \n",
       "Corynebacter_diphtheria_Cas9          0  \n",
       "F__novicida_Cas9                      0  \n",
       "Lactobacillus_buchneri_Cas9          23  \n",
       "Listeria_innocua_Cas9               149  \n",
       "N__meningitidis_Cas9                  2  \n",
       "Neisseria_cinerea_Cas9                2  \n",
       "P__multocida_Cas9                     3  \n",
       "Parvibaculum_lavamentivorans_Cas9     0  \n",
       "S_CRISPR_1_thermophilus_Cas9          1  \n",
       "S_CRISPR_3_thermophilus_Cas9        374  \n",
       "S__mutans_Cas9                      316  \n",
       "S__pyogenes_Cas9                     68  \n",
       "Staphylococcus_aureus_Cas9            6  \n",
       "Streptococcus_pasteurianus_Cas9       2  \n",
       "T__denticola_Cas9                    42  "
      ]
     },
     "execution_count": 629,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_dfs[97].loc['S__pyogenes_Cas9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 630,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>Num High AA</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C__jejuni_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Campylobacter_lari_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Corynebacter_diphtheria_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F__novicida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lactobacillus_buchneri_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Listeria_innocua_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>41</td>\n",
       "      <td>23</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N__meningitidis_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neisseria_cinerea_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P__multocida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Parvibaculum_lavamentivorans_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_CRISPR_1_thermophilus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_CRISPR_3_thermophilus_Cas9</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>170</td>\n",
       "      <td>101</td>\n",
       "      <td>59</td>\n",
       "      <td>30</td>\n",
       "      <td>13</td>\n",
       "      <td>86</td>\n",
       "      <td>375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__mutans_Cas9</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>131</td>\n",
       "      <td>91</td>\n",
       "      <td>59</td>\n",
       "      <td>33</td>\n",
       "      <td>15</td>\n",
       "      <td>124</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__pyogenes_Cas9</th>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>130</td>\n",
       "      <td>104</td>\n",
       "      <td>79</td>\n",
       "      <td>54</td>\n",
       "      <td>31</td>\n",
       "      <td>408</td>\n",
       "      <td>408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Staphylococcus_aureus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus_pasteurianus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T__denticola_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   10   11    5    6   7   8   9 Num High AA  \\\n",
       "C__jejuni_Cas9                      0    0    0    0   0   0   0           0   \n",
       "Campylobacter_lari_Cas9             0    0    0    0   0   0   0           0   \n",
       "Corynebacter_diphtheria_Cas9        0    0    0    0   0   0   0           0   \n",
       "F__novicida_Cas9                    0    0    0    0   0   0   0           0   \n",
       "Lactobacillus_buchneri_Cas9         0    0   15    5   2   1   0           4   \n",
       "Listeria_innocua_Cas9               0    0   67   41  23  13   5           8   \n",
       "N__meningitidis_Cas9                0    0    2    0   0   0   0           0   \n",
       "Neisseria_cinerea_Cas9              0    0    2    0   0   0   0           0   \n",
       "P__multocida_Cas9                   0    0    2    1   0   0   0           0   \n",
       "Parvibaculum_lavamentivorans_Cas9   0    0    0    0   0   0   0           0   \n",
       "S_CRISPR_1_thermophilus_Cas9        0    0    1    0   0   0   0           0   \n",
       "S_CRISPR_3_thermophilus_Cas9        2    0  170  101  59  30  13          86   \n",
       "S__mutans_Cas9                      2    0  131   91  59  33  15         124   \n",
       "S__pyogenes_Cas9                   10  NaN  130  104  79  54  31         408   \n",
       "Staphylococcus_aureus_Cas9          0    0    6    0   0   0   0           0   \n",
       "Streptococcus_pasteurianus_Cas9     0    0    2    0   0   0   0           1   \n",
       "T__denticola_Cas9                   0    0   27   12   2   1   0           3   \n",
       "\n",
       "                                  Total  \n",
       "C__jejuni_Cas9                        0  \n",
       "Campylobacter_lari_Cas9               0  \n",
       "Corynebacter_diphtheria_Cas9          0  \n",
       "F__novicida_Cas9                      0  \n",
       "Lactobacillus_buchneri_Cas9          23  \n",
       "Listeria_innocua_Cas9               149  \n",
       "N__meningitidis_Cas9                  2  \n",
       "Neisseria_cinerea_Cas9                2  \n",
       "P__multocida_Cas9                     3  \n",
       "Parvibaculum_lavamentivorans_Cas9     0  \n",
       "S_CRISPR_1_thermophilus_Cas9          1  \n",
       "S_CRISPR_3_thermophilus_Cas9        375  \n",
       "S__mutans_Cas9                      331  \n",
       "S__pyogenes_Cas9                    408  \n",
       "Staphylococcus_aureus_Cas9            6  \n",
       "Streptococcus_pasteurianus_Cas9       2  \n",
       "T__denticola_Cas9                    42  "
      ]
     },
     "execution_count": 630,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_dfs[74].loc['S__pyogenes_Cas9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/carlomazzaferro/Documents/Code/neoantigen/antigen_discovery/nepitope/pairwise_comp.py:193: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  df['Peptide'] = df['Peptide'].str.replace('X', '-')\n"
     ]
    }
   ],
   "source": [
    "pwcomp_1_swap = pairwise_comp.PairwiseComp(ls_mod, 5, mod_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_1_swap = pwcomp_1_swap.pipe_run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_1_swap.to_csv('/Users/carlomazzaferro/Desktop/Test_IEDB/OtherProtsPrashant/single_swap_pairwise_comp.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>Matches Loc</th>\n",
       "      <th>Num High AA</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C__jejuni_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Campylobacter_lari_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Corynebacter_diphtheria_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F__novicida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lactobacillus_buchneri_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[1266, 1267, 1268, 1269, 1270], [1266, 1267,...</td>\n",
       "      <td>5</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Listeria_innocua_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>66</td>\n",
       "      <td>40</td>\n",
       "      <td>23</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>118</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N__meningitidis_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neisseria_cinerea_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P__multocida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Parvibaculum_lavamentivorans_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__mutans_Cas9</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>119</td>\n",
       "      <td>81</td>\n",
       "      <td>49</td>\n",
       "      <td>24</td>\n",
       "      <td>10</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>202</td>\n",
       "      <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__pyogenes_Cas9</th>\n",
       "      <td>48</td>\n",
       "      <td>3</td>\n",
       "      <td>513</td>\n",
       "      <td>415</td>\n",
       "      <td>320</td>\n",
       "      <td>227</td>\n",
       "      <td>136</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>1659</td>\n",
       "      <td>1662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__thermophilus_CRISPR_3_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>160</td>\n",
       "      <td>93</td>\n",
       "      <td>52</td>\n",
       "      <td>23</td>\n",
       "      <td>8</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>210</td>\n",
       "      <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Staphylococcus_aureus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11], [650, ...</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus_pasteurianus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [1270, 1271, 1272, 1273, ...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T__denticola_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[164, 165, 166, 167, 168], [810, 811, 812, 8...</td>\n",
       "      <td>41</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   10 11    5    6    7    8    9  \\\n",
       "C__jejuni_Cas9                      0  0    0    0    0    0    0   \n",
       "Campylobacter_lari_Cas9             0  0    0    0    0    0    0   \n",
       "Corynebacter_diphtheria_Cas9        0  0    0    0    0    0    0   \n",
       "F__novicida_Cas9                    0  0    0    0    0    0    0   \n",
       "Lactobacillus_buchneri_Cas9         0  0   13    5    2    1    0   \n",
       "Listeria_innocua_Cas9               0  0   66   40   23   13    5   \n",
       "N__meningitidis_Cas9                0  0    2    0    0    0    0   \n",
       "Neisseria_cinerea_Cas9              0  0    2    0    0    0    0   \n",
       "P__multocida_Cas9                   0  0    2    1    0    0    0   \n",
       "Parvibaculum_lavamentivorans_Cas9   0  0    0    0    0    0    0   \n",
       "S__mutans_Cas9                      1  0  119   81   49   24   10   \n",
       "S__pyogenes_Cas9                   48  3  513  415  320  227  136   \n",
       "S__thermophilus_CRISPR_3_Cas9       0  0  160   93   52   23    8   \n",
       "Staphylococcus_aureus_Cas9          0  0    6    0    0    0    0   \n",
       "Streptococcus_pasteurianus_Cas9     0  0    2    0    0    0    0   \n",
       "T__denticola_Cas9                   0  0   27   12    2    1    0   \n",
       "\n",
       "                                                                         Matches Loc  \\\n",
       "C__jejuni_Cas9                                                                  [[]]   \n",
       "Campylobacter_lari_Cas9                                                         [[]]   \n",
       "Corynebacter_diphtheria_Cas9                                                    [[]]   \n",
       "F__novicida_Cas9                                                                [[]]   \n",
       "Lactobacillus_buchneri_Cas9        [[[1266, 1267, 1268, 1269, 1270], [1266, 1267,...   \n",
       "Listeria_innocua_Cas9              [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "N__meningitidis_Cas9                        [[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]   \n",
       "Neisseria_cinerea_Cas9                      [[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]   \n",
       "P__multocida_Cas9                                                               [[]]   \n",
       "Parvibaculum_lavamentivorans_Cas9                                               [[]]   \n",
       "S__mutans_Cas9                     [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "S__pyogenes_Cas9                   [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "S__thermophilus_CRISPR_3_Cas9      [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "Staphylococcus_aureus_Cas9         [[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11], [650, ...   \n",
       "Streptococcus_pasteurianus_Cas9    [[[7, 8, 9, 10, 11], [1270, 1271, 1272, 1273, ...   \n",
       "T__denticola_Cas9                  [[[164, 165, 166, 167, 168], [810, 811, 812, 8...   \n",
       "\n",
       "                                  Num High AA Total  \n",
       "C__jejuni_Cas9                              0     0  \n",
       "Campylobacter_lari_Cas9                     0     0  \n",
       "Corynebacter_diphtheria_Cas9                0     0  \n",
       "F__novicida_Cas9                            0     0  \n",
       "Lactobacillus_buchneri_Cas9                 5    21  \n",
       "Listeria_innocua_Cas9                     118   147  \n",
       "N__meningitidis_Cas9                        2     2  \n",
       "Neisseria_cinerea_Cas9                      2     2  \n",
       "P__multocida_Cas9                           0     3  \n",
       "Parvibaculum_lavamentivorans_Cas9           0     0  \n",
       "S__mutans_Cas9                            202   284  \n",
       "S__pyogenes_Cas9                         1659  1662  \n",
       "S__thermophilus_CRISPR_3_Cas9             210   336  \n",
       "Staphylococcus_aureus_Cas9                  6     6  \n",
       "Streptococcus_pasteurianus_Cas9             2     2  \n",
       "T__denticola_Cas9                          41    42  "
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_1_swap.loc['S__pyogenes_Cas9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>Matches Loc</th>\n",
       "      <th>Num High AA</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C__jejuni_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Campylobacter_lari_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Corynebacter_diphtheria_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F__novicida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lactobacillus_buchneri_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[1266, 1267, 1268, 1269, 1270], [1266, 1267,...</td>\n",
       "      <td>7</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Listeria_innocua_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>41</td>\n",
       "      <td>23</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>120</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N__meningitidis_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neisseria_cinerea_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P__multocida_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Parvibaculum_lavamentivorans_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[]]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__mutans_Cas9</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>131</td>\n",
       "      <td>91</td>\n",
       "      <td>59</td>\n",
       "      <td>33</td>\n",
       "      <td>15</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>249</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__pyogenes_Cas9</th>\n",
       "      <td>52</td>\n",
       "      <td>3</td>\n",
       "      <td>542</td>\n",
       "      <td>441</td>\n",
       "      <td>343</td>\n",
       "      <td>245</td>\n",
       "      <td>147</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>1770</td>\n",
       "      <td>1773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S__thermophilus_CRISPR_3_Cas9</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>171</td>\n",
       "      <td>101</td>\n",
       "      <td>59</td>\n",
       "      <td>30</td>\n",
       "      <td>13</td>\n",
       "      <td>[[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...</td>\n",
       "      <td>250</td>\n",
       "      <td>376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Staphylococcus_aureus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11], [650, ...</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus_pasteurianus_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[7, 8, 9, 10, 11], [1270, 1271, 1272, 1273, ...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T__denticola_Cas9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>[[[164, 165, 166, 167, 168], [810, 811, 812, 8...</td>\n",
       "      <td>41</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   10 11    5    6    7    8    9  \\\n",
       "C__jejuni_Cas9                      0  0    0    0    0    0    0   \n",
       "Campylobacter_lari_Cas9             0  0    0    0    0    0    0   \n",
       "Corynebacter_diphtheria_Cas9        0  0    0    0    0    0    0   \n",
       "F__novicida_Cas9                    0  0    0    0    0    0    0   \n",
       "Lactobacillus_buchneri_Cas9         0  0   15    5    2    1    0   \n",
       "Listeria_innocua_Cas9               0  0   67   41   23   13    5   \n",
       "N__meningitidis_Cas9                0  0    2    0    0    0    0   \n",
       "Neisseria_cinerea_Cas9              0  0    2    0    0    0    0   \n",
       "P__multocida_Cas9                   0  0    2    1    0    0    0   \n",
       "Parvibaculum_lavamentivorans_Cas9   0  0    0    0    0    0    0   \n",
       "S__mutans_Cas9                      2  0  131   91   59   33   15   \n",
       "S__pyogenes_Cas9                   52  3  542  441  343  245  147   \n",
       "S__thermophilus_CRISPR_3_Cas9       2  0  171  101   59   30   13   \n",
       "Staphylococcus_aureus_Cas9          0  0    6    0    0    0    0   \n",
       "Streptococcus_pasteurianus_Cas9     0  0    2    0    0    0    0   \n",
       "T__denticola_Cas9                   0  0   27   12    2    1    0   \n",
       "\n",
       "                                                                         Matches Loc  \\\n",
       "C__jejuni_Cas9                                                                  [[]]   \n",
       "Campylobacter_lari_Cas9                                                         [[]]   \n",
       "Corynebacter_diphtheria_Cas9                                                    [[]]   \n",
       "F__novicida_Cas9                                                                [[]]   \n",
       "Lactobacillus_buchneri_Cas9        [[[1266, 1267, 1268, 1269, 1270], [1266, 1267,...   \n",
       "Listeria_innocua_Cas9              [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "N__meningitidis_Cas9                        [[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]   \n",
       "Neisseria_cinerea_Cas9                      [[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11]]]   \n",
       "P__multocida_Cas9                                                               [[]]   \n",
       "Parvibaculum_lavamentivorans_Cas9                                               [[]]   \n",
       "S__mutans_Cas9                     [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "S__pyogenes_Cas9                   [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "S__thermophilus_CRISPR_3_Cas9      [[[8, 9, 10, 11, 12, 13, 14, 15], [7, 8, 9, 10...   \n",
       "Staphylococcus_aureus_Cas9         [[[7, 8, 9, 10, 11], [7, 8, 9, 10, 11], [650, ...   \n",
       "Streptococcus_pasteurianus_Cas9    [[[7, 8, 9, 10, 11], [1270, 1271, 1272, 1273, ...   \n",
       "T__denticola_Cas9                  [[[164, 165, 166, 167, 168], [810, 811, 812, 8...   \n",
       "\n",
       "                                  Num High AA Total  \n",
       "C__jejuni_Cas9                              0     0  \n",
       "Campylobacter_lari_Cas9                     0     0  \n",
       "Corynebacter_diphtheria_Cas9                0     0  \n",
       "F__novicida_Cas9                            0     0  \n",
       "Lactobacillus_buchneri_Cas9                 7    23  \n",
       "Listeria_innocua_Cas9                     120   149  \n",
       "N__meningitidis_Cas9                        2     2  \n",
       "Neisseria_cinerea_Cas9                      2     2  \n",
       "P__multocida_Cas9                           0     3  \n",
       "Parvibaculum_lavamentivorans_Cas9           0     0  \n",
       "S__mutans_Cas9                            249   331  \n",
       "S__pyogenes_Cas9                         1770  1773  \n",
       "S__thermophilus_CRISPR_3_Cas9             250   376  \n",
       "Staphylococcus_aureus_Cas9                  6     6  \n",
       "Streptococcus_pasteurianus_Cas9             2     2  \n",
       "T__denticola_Cas9                          41    42  "
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pairwise_comps.loc['S__pyogenes_Cas9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [py35]",
   "language": "python",
   "name": "Python [py35]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
